// @primer/react/lib-esm/Token/AvatarToken.js
import { c } from 'react-compiler-runtime';
import { forwardRef } from 'react';
import { defaultTokenSize, tokenSizes } from './TokenBase.js';
import Token from './Token.js';
import classes from './AvatarToken.module.css.js';
import { clsx } from 'clsx';
import { jsx } from 'react/jsx-runtime';
import Avatar from '../Avatar/Avatar.js';

const AvatarToken = /*#__PURE__*/forwardRef((t0, forwardedRef) => {
  const $ = c(20);
  let avatarSrc;
  let className;
  let id;
  let rest;
  let t1;
  if ($[0] !== t0) {
    ({
      avatarSrc,
      id,
      size: t1,
      className,
      ...rest
    } = t0);
    $[0] = t0;
    $[1] = avatarSrc;
    $[2] = className;
    $[3] = id;
    $[4] = rest;
    $[5] = t1;
  } else {
    avatarSrc = $[1];
    className = $[2];
    id = $[3];
    rest = $[4];
    t1 = $[5];
  }
  const size = t1 === undefined ? defaultTokenSize : t1;
  let t2;
  if ($[6] !== avatarSrc || $[7] !== size) {
    t2 = () => /*#__PURE__*/jsx("span", {
      className: classes.AvatarContainer,
      "data-size": size,
      children: /*#__PURE__*/jsx(Avatar, {
        src: avatarSrc,
        size: parseInt(tokenSizes[size], 10),
        className: classes.Avatar
      })
    });
    $[6] = avatarSrc;
    $[7] = size;
    $[8] = t2;
  } else {
    t2 = $[8];
  }
  let t3;
  if ($[9] !== id) {
    var _id;
    t3 = (_id = id) === null || _id === void 0 ? void 0 : _id.toString();
    $[9] = id;
    $[10] = t3;
  } else {
    t3 = $[10];
  }
  let t4;
  if ($[11] !== className) {
    t4 = clsx(classes.Token, className);
    $[11] = className;
    $[12] = t4;
  } else {
    t4 = $[12];
  }
  let t5;
  if ($[13] !== forwardedRef || $[14] !== rest || $[15] !== size || $[16] !== t2 || $[17] !== t3 || $[18] !== t4) {
    t5 = /*#__PURE__*/jsx(Token, {
      leadingVisual: t2,
      size: size,
      id: t3,
      className: t4,
      ...rest,
      ref: forwardedRef
    });
    $[13] = forwardedRef;
    $[14] = rest;
    $[15] = size;
    $[16] = t2;
    $[17] = t3;
    $[18] = t4;
    $[19] = t5;
  } else {
    t5 = $[19];
  }
  return t5;
});
AvatarToken.displayName = 'AvatarToken';

export { AvatarToken as default };
