// @sprucelabs/spruce-event-utils/build/.spruce/schemas/spruceEventUtils/v2021_09_13/eventTarget.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const eventTargetSchema = {
    id: 'eventTarget',
    version: 'v2021_09_13',
    namespace: 'SpruceEventUtils',
    name: 'event target',
    importsWhenRemote: ['import \'@sprucelabs/spruce-event-utils\'',],
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-event-utils',
    fields: {
        /** . */
        'locationId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'personId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'organizationId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'skillId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'roleId': {
            type: 'id',
            options: undefined
        },
    }
};
schema_1.SchemaRegistry.getInstance().trackSchema(eventTargetSchema);
exports.default = eventTargetSchema;
