// @sprucelabs/spruce-event-utils/build/.spruce/schemas/spruceEventUtils/v2021_09_13/eventPagingRequest.schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const eventPagingSort_schema_1 = __importDefault(require("./eventPagingSort.schema"));
const eventPagingRequestSchema = {
    id: 'eventPagingRequest',
    version: 'v2021_09_13',
    namespace: 'SpruceEventUtils',
    name: 'Event paging request',
    importsWhenRemote: ['import \'@sprucelabs/spruce-event-utils\'',],
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-event-utils',
    fields: {
        /** . */
        'pageSize': {
            type: 'number',
            options: undefined
        },
        /** . */
        'next': {
            type: 'id',
            options: undefined
        },
        /** . */
        'previous': {
            type: 'id',
            options: undefined
        },
        /** . */
        'sort': {
            type: 'schema',
            isArray: true,
            options: { schema: eventPagingSort_schema_1.default, }
        },
    }
};
schema_1.SchemaRegistry.getInstance().trackSchema(eventPagingRequestSchema);
exports.default = eventPagingRequestSchema;
