// circuit-json-to-simple-3d/dist/index.js
// lib/index.ts
import { cju } from "@tscircuit/circuit-json-util";
import { renderScene } from "@tscircuit/simple-3d-svg";

// lib/getDefaultCameraForPcbBoard.ts
function getDefaultCameraForPcbBoard(pcbBoard) {
  const w = pcbBoard.width;
  const h = pcbBoard.height;
  const cx = pcbBoard.center?.x;
  const cz = pcbBoard.center?.y;
  const dist = Math.max(w, h) * 1.5;
  return {
    position: { x: cx - dist, y: dist, z: cz - dist },
    lookAt: { x: cx, y: 0, z: cz },
    focalLength: 2
  };
}

// lib/index.ts
function convertCircuitJsonToSimple3dSvg(circuitJson, opts = {}) {
  const db = cju(circuitJson);
  const boxes = [];
  const pcbBoard = db.pcb_board.list()[0];
  if (!pcbBoard) throw new Error("No pcb_board, can't render to 3d");
  const camera = opts.camera ?? getDefaultCameraForPcbBoard(pcbBoard);
  if (!camera.focalLength) {
    camera.focalLength = 1;
  }
  boxes.push({
    center: {
      x: pcbBoard.center.x,
      y: 0,
      z: pcbBoard.center.y
    },
    size: {
      x: pcbBoard.width,
      y: pcbBoard.thickness,
      z: pcbBoard.height
    },
    color: "rgba(0,140,0,0.8)"
  });
  const DEFAULT_COMP_HEIGHT = 2;
  for (const comp of db.pcb_component.list()) {
    const sourceComponent = db.source_component.get(comp.source_component_id);
    boxes.push({
      center: {
        x: comp.center.x,
        y: pcbBoard.thickness / 2 + DEFAULT_COMP_HEIGHT / 2,
        z: comp.center.y
      },
      size: {
        x: comp.width,
        y: Math.min(Math.min(comp.width, comp.height), DEFAULT_COMP_HEIGHT),
        z: comp.height
      },
      color: "rgba(128,128,128,0.9)",
      topLabel: sourceComponent?.name ?? "?",
      topLabelColor: "white"
    });
  }
  return renderScene(
    { boxes, camera },
    {
      backgroundColor: "lightgray"
    }
  );
}
export {
  convertCircuitJsonToSimple3dSvg
};
