// @sprucelabs/sprucebot-llm/build/bots/PromptGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setUndefinedToNull = setUndefinedToNull;
const schema_1 = require("@sprucelabs/schema");
const eta_1 = require("eta");
const templates_1 = require("./templates");
class PromptGenerator {
    constructor(bot, options) {
        this.eta = new eta_1.Eta();
        this.log = process.env.SHOULD_LOG_GENERATED_PROMPTS === 'true'
            ? console.info
            : () => { };
        (0, schema_1.assertOptions)({ bot }, ['bot']);
        this.bot = bot;
        this.promptTemplate = options?.promptTemplate ?? templates_1.PROMPT_TEMPLATE;
    }
    static Generator(bot, options) {
        return new (this.Class ?? PromptGenerator)(bot, options);
    }
    async generate() {
        const { stateSchema, state, ...rest } = this.bot.serialize();
        const { stateSchemaJson, stateJson } = this.stringifyState(stateSchema, state);
        const rendered = await this.eta.renderStringAsync(this.promptTemplate, {
            stateSchemaJson,
            stateJson,
            ...rest,
        });
        this.log('Generated prompt:', rendered);
        return rendered;
    }
    stringifyState(stateSchema, state) {
        if (!stateSchema) {
            return {
                stateSchemaJson: null,
                stateJson: null,
            };
        }
        const normalizedState = (0, schema_1.normalizeSchemaValues)(stateSchema, state ?? {}, {});
        setUndefinedToNull(normalizedState);
        const stateSchemaJson = JSON.stringify(stateSchema);
        const stateJson = JSON.stringify(normalizedState);
        return { stateSchemaJson, stateJson };
    }
}
exports.default = PromptGenerator;
function setUndefinedToNull(obj) {
    for (const key in obj) {
        if (obj.hasOwnProperty(key)) {
            if (obj[key] === undefined) {
                obj[key] = null;
            }
            else if (typeof obj[key] === 'object') {
                setUndefinedToNull(obj[key]);
            }
        }
    }
}
