// @sprucelabs/sprucebot-llm/build/bots/SprucebotLlmSkillImpl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mercury_event_emitter_1 = require("@sprucelabs/mercury-event-emitter");
const schema_1 = require("@sprucelabs/schema");
const llm_types_1 = require("../llm.types");
class SprucebotLlmSkillImpl extends mercury_event_emitter_1.AbstractEventEmitter {
    constructor(options) {
        super(llm_types_1.llmEventContract);
        this.state = {};
        const { state, stateSchema, ...rest } = options ?? {};
        this.options = { ...rest, stateSchema };
        this.stateSchema = stateSchema;
        this.state = stateSchema
            ? {
                ...(0, schema_1.defaultSchemaValues)(stateSchema),
                ...state,
            }
            : undefined;
    }
    async updateState(updates) {
        await this.emit('did-update-state');
        this.state = { ...this.state, ...updates };
    }
    getState() {
        return this.state;
    }
    serialize() {
        return {
            ...this.options,
            stateSchema: this.stateSchema,
            state: this.state,
        };
    }
}
exports.default = SprucebotLlmSkillImpl;
