// @sprucelabs/mercury-types/build/.spruce/schemas/mercury/v2020_12_25/eventContract.schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const eventSignaturesByName_schema_1 = __importDefault(require("./eventSignaturesByName.schema"));
const eventContractSchema = {
    id: 'eventContract',
    version: 'v2020_12_25',
    namespace: 'Mercury',
    name: 'Event contract',
    importsWhenRemote: ['import \'@sprucelabs/mercury-types\'',],
    moduleToImportFromWhenRemote: '@sprucelabs/mercury-types',
    description: 'A Mercury Event Contract.',
    fields: {
        /** . */
        'id': {
            type: 'id',
            options: undefined
        },
        /** . */
        'eventSignatures': {
            type: 'schema',
            isRequired: true,
            options: { schema: eventSignaturesByName_schema_1.default, }
        },
    }
};
schema_1.SchemaRegistry.getInstance().trackSchema(eventContractSchema);
exports.default = eventContractSchema;
