// @sprucelabs/mercury-types/build/.spruce/schemas/mercury/v2020_12_25/permission.schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const permissionDefaults_schema_1 = __importDefault(require("./permissionDefaults.schema"));
const statusFlags_schema_1 = __importDefault(require("./statusFlags.schema"));
const permissionSchema = {
    id: 'permission',
    version: 'v2020_12_25',
    namespace: 'Mercury',
    name: 'Permission',
    moduleToImportFromWhenRemote: '@sprucelabs/mercury-types',
    fields: {
        /** id. Hyphen separated id for this permission, e.g. can-unlock-doors */
        'id': {
            label: 'id',
            type: 'text',
            isRequired: true,
            hint: 'Hyphen separated id for this permission, e.g. can-unlock-doors',
            options: undefined
        },
        /** Name. Human readable name for this permission */
        'name': {
            label: 'Name',
            type: 'text',
            isRequired: true,
            hint: 'Human readable name for this permission',
            options: undefined
        },
        /** Description. */
        'description': {
            label: 'Description',
            type: 'text',
            options: undefined
        },
        /** Require all statuses. */
        'requireAllStatuses': {
            label: 'Require all statuses',
            type: 'boolean',
            defaultValue: false,
            options: undefined
        },
        /** . */
        'defaults': {
            type: 'schema',
            options: { schema: permissionDefaults_schema_1.default, }
        },
        /** . */
        'can': {
            type: 'schema',
            options: { schema: statusFlags_schema_1.default, }
        },
    }
};
schema_1.SchemaRegistry.getInstance().trackSchema(permissionSchema);
exports.default = permissionSchema;
