// @sprucelabs/mercury-types/build/.spruce/schemas/mercury/v2020_12_25/permissionReference.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const permissionReferenceSchema = {
    id: 'permissionReference',
    version: 'v2020_12_25',
    namespace: 'Mercury',
    name: '',
    importsWhenRemote: ['import * as MercuryTypes from \'@sprucelabs/mercury-types\'',],
    moduleToImportFromWhenRemote: '@sprucelabs/mercury-types',
    typeSuffix: '<ContractId extends MercuryTypes.PermissionContractId = MercuryTypes.PermissionContractId>',
    fields: {
        /** . */
        'contractId': {
            type: 'raw',
            isRequired: true,
            options: { valueType: `ContractId`, }
        },
        /** . */
        'permissionIdsAny': {
            type: 'raw',
            isArray: true,
            options: { valueType: `MercuryTypes.PermissionId<ContractId>`, }
        },
    }
};
schema_1.SchemaRegistry.getInstance().trackSchema(permissionReferenceSchema);
exports.default = permissionReferenceSchema;
