// @sprucelabs/mercury-types/build/.spruce/schemas/mercury/v2020_12_25/permissionContract.schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const permission_schema_1 = __importDefault(require("./permission.schema"));
const permissionContractSchema = {
    id: 'permissionContract',
    version: 'v2020_12_25',
    namespace: 'Mercury',
    name: 'Permission contract',
    importsWhenRemote: ['import \'@sprucelabs/mercury-types\'',],
    moduleToImportFromWhenRemote: '@sprucelabs/mercury-types',
    description: 'A Mercury Permission Contract',
    fields: {
        /** . */
        'id': {
            type: 'text',
            isRequired: true,
            options: undefined
        },
        /** Name. Human readable name for this contract */
        'name': {
            label: 'Name',
            type: 'text',
            isRequired: true,
            hint: 'Human readable name for this contract',
            options: undefined
        },
        /** Description. */
        'description': {
            label: 'Description',
            type: 'text',
            options: undefined
        },
        /** Require all permissions. */
        'requireAllPermissions': {
            label: 'Require all permissions',
            type: 'boolean',
            defaultValue: false,
            options: undefined
        },
        /** . */
        'permissions': {
            type: 'schema',
            isRequired: true,
            isArray: true,
            minArrayLength: 0,
            options: { schema: permission_schema_1.default, }
        },
    }
};
schema_1.SchemaRegistry.getInstance().trackSchema(permissionContractSchema);
exports.default = permissionContractSchema;
