// @sprucelabs/spruce-core-schemas/build/schemas/v2020_07_22/message.builder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageSourceSchema = exports.messageTargetSchema = void 0;
const schema_1 = require("@sprucelabs/schema");
const choice_builder_1 = __importDefault(require("./choice.builder"));
const link_builder_1 = __importDefault(require("./link.builder"));
const eventTargetFields = {
    locationId: {
        type: 'id',
    },
    personId: {
        type: 'id',
    },
    organizationId: {
        type: 'id',
    },
    skillId: {
        type: 'id',
    },
    roleId: {
        type: 'id',
    },
};
exports.messageTargetSchema = (0, schema_1.buildSchema)({
    id: 'messageTarget',
    fields: {
        ...eventTargetFields,
        phone: {
            type: 'phone',
        },
    },
});
exports.messageSourceSchema = (0, schema_1.buildSchema)({
    id: 'messageSource',
    fields: {
        ...eventTargetFields,
        isCore: {
            type: 'boolean',
        },
    },
});
exports.default = (0, schema_1.buildSchema)({
    id: 'message',
    description: 'A discrete communication between two humans or a human and a machine.',
    fields: {
        id: {
            type: 'id',
            isRequired: true,
        },
        trackingId: {
            type: 'id',
            hint: 'An arbitrary id that can be used to track this message when being sent and not yet assigned an Id.',
        },
        dateCreated: {
            type: 'number',
            isRequired: true,
        },
        dateSent: {
            type: 'number',
        },
        target: {
            type: 'schema',
            isRequired: true,
            options: {
                schema: exports.messageTargetSchema,
            },
        },
        source: {
            type: 'schema',
            isRequired: true,
            options: {
                schema: exports.messageSourceSchema,
            },
        },
        errors: {
            type: 'text',
            isArray: true,
            isPrivate: true,
        },
        classification: {
            type: 'select',
            isRequired: true,
            options: {
                choices: [
                    {
                        value: 'auth',
                        label: 'Auth',
                    },
                    {
                        value: 'transactional',
                        label: 'transactional',
                    },
                    {
                        value: 'promotional',
                        label: 'Promotional',
                    },
                    {
                        value: 'incoming',
                        label: 'incoming',
                    },
                ],
            },
        },
        status: {
            type: 'select',
            defaultValue: 'pending',
            isPrivate: true,
            options: {
                choices: [
                    {
                        value: 'pending',
                        label: 'Pending',
                    },
                    {
                        value: 'processing',
                        label: 'Processing',
                    },
                    {
                        value: 'sent',
                        label: 'Sent',
                    },
                    {
                        value: 'failed',
                        label: 'Failed',
                    },
                    {
                        value: 'ignored',
                        label: 'Ignored',
                    },
                ],
            },
        },
        body: {
            type: 'text',
            isRequired: true,
        },
        context: {
            type: 'raw',
            options: {
                valueType: 'Record<string, any>',
            },
            isPrivate: true,
        },
        topicId: {
            type: 'id',
        },
        links: {
            type: 'schema',
            isArray: true,
            minArrayLength: 0,
            options: {
                schema: link_builder_1.default,
            },
        },
        choices: {
            type: 'schema',
            isArray: true,
            options: {
                schema: choice_builder_1.default,
            },
        },
    },
});
