// @sprucelabs/spruce-core-schemas/build/schemas/v2020_07_22/organization.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
exports.default = (0, schema_1.buildSchema)({
    id: 'organization',
    name: 'Organization',
    description: 'A company or team. Comprises of many people and locations.',
    fields: {
        id: {
            label: 'Id',
            type: 'id',
            isRequired: true,
        },
        name: {
            label: 'Name',
            type: 'text',
            isRequired: true,
        },
        address: {
            label: 'Address',
            type: 'address',
        },
        slug: {
            label: 'Slug',
            type: 'text',
            isRequired: true,
        },
        isPublic: {
            label: 'Public',
            type: 'boolean',
            hint: 'Is this organization viewable by guests?',
            defaultValue: false,
        },
        dateCreated: {
            type: 'dateTime',
            isRequired: true,
        },
        dateDeleted: {
            type: 'dateTime',
        },
    },
});
