// @sprucelabs/spruce-core-schemas/build/schemas/v2020_07_22/person.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const timezoneChoices_1 = require("../../timezoneChoices");
exports.default = (0, schema_1.buildSchema)({
    id: 'person',
    name: 'Person',
    description: 'A human being.',
    fields: {
        id: {
            label: 'Id',
            type: 'id',
            isRequired: true,
        },
        firstName: {
            label: 'First name',
            type: 'text',
            isPrivate: true,
        },
        lastName: {
            label: 'Last name',
            type: 'text',
            isPrivate: true,
        },
        casualName: {
            label: 'Casual name',
            type: 'text',
            hint: 'The name you can use when talking to this person.',
            isRequired: true,
        },
        timezone: {
            label: 'Timezone',
            type: 'select',
            options: {
                choices: timezoneChoices_1.timezoneChoices,
            },
        },
        phone: {
            label: 'Phone',
            type: 'phone',
            hint: 'A number that can be texted',
            isPrivate: true,
        },
        username: {
            label: 'Username',
            type: 'text',
            hint: 'An optional username if the person does not want to login using their phone',
            isPrivate: true,
        },
        avatar: {
            label: 'Avatar src',
            type: 'image',
            options: {
                requiredSizes: ['*'],
            },
        },
        dateCreated: {
            type: 'dateTime',
            isRequired: true,
        },
        dateUpdated: {
            type: 'dateTime',
        },
        dateScrambled: {
            type: 'dateTime',
        },
    },
});
