// @sprucelabs/schema/build/fields/PhoneField.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const formatPhoneNumber_1 = __importStar(require("../utilities/formatPhoneNumber"));
const AbstractField_1 = __importDefault(require("./AbstractField"));
class PhoneField extends AbstractField_1.default {
    static generateTemplateDetails(options) {
        return {
            valueType: `string${options.definition.isArray ? '[]' : ''}`,
        };
    }
    toValueType(value) {
        const stringValue = `${value}`;
        const phoneNumber = (0, formatPhoneNumber_1.default)(stringValue);
        return phoneNumber;
    }
    validate(value, options) {
        const errors = super.validate(value, options);
        if (errors.length === 0 && this.isRequired && `${value}`.length === 0) {
            errors.push({
                code: 'MISSING_PARAMETER',
                label: this.label,
                name: this.name,
            });
        }
        else if (errors.length === 0 && value && !(0, formatPhoneNumber_1.isValidNumber)(value)) {
            errors.push({
                code: 'INVALID_PARAMETER',
                name: this.name,
                label: this.label,
                friendlyMessage: `Please enter a valid phone number.`,
            });
        }
        return errors;
    }
}
PhoneField.description = 'Takes anything close to a phone number and formats it. Also great at validating numbers.';
exports.default = PhoneField;
