// @primer/react/lib/FormControl/FormControlCaption.js
'use strict';

var reactCompilerRuntime = require('react-compiler-runtime');
var clsx = require('clsx');
var FormControlCaption_module = require('./FormControlCaption.module.css.js');
var _FormControlContext = require('./_FormControlContext.js');
var BoxWithFallback = require('../internal/components/BoxWithFallback.js');
var jsxRuntime = require('react/jsx-runtime');
var Text = require('../Text/Text.js');

function FormControlCaption(t0) {
  const $ = reactCompilerRuntime.c(8);
  const {
    id,
    children,
    sx,
    className
  } = t0;
  const {
    captionId,
    disabled
  } = _FormControlContext.useFormControlContext();
  const t1 = id !== null && id !== void 0 ? id : captionId;
  let t2;
  if ($[0] !== className) {
    t2 = clsx.clsx(className, FormControlCaption_module.Caption);
    $[0] = className;
    $[1] = t2;
  } else {
    t2 = $[1];
  }
  const t3 = disabled ? "" : undefined;
  let t4;
  if ($[2] !== children || $[3] !== sx || $[4] !== t1 || $[5] !== t2 || $[6] !== t3) {
    t4 = /*#__PURE__*/jsxRuntime.jsx(BoxWithFallback.BoxWithFallback, {
      as: Text,
      id: t1,
      className: t2,
      "data-control-disabled": t3,
      sx: sx,
      children: children
    });
    $[2] = children;
    $[3] = sx;
    $[4] = t1;
    $[5] = t2;
    $[6] = t3;
    $[7] = t4;
  } else {
    t4 = $[7];
  }
  return t4;
}

exports.FormControlCaption = FormControlCaption;
