// @tscircuit/math-utils/dist/index.js
import {
  pointToSegmentClosestPoint,
  segmentToBoundsMinDistance,
  segmentToBoxMinDistance,
  segmentToCircleMinDistance,
  segmentToSegmentMinDistance
} from "./chunk-FWQGMQBW.js";
import "./chunk-GYQ2KZV6.js";
import {
  getUnitVectorFromDirection,
  getUnitVectorFromPointAToB
} from "./chunk-GIGMPRPV.js";
import {
  grid
} from "./chunk-U45EKA3R.js";
import {
  distSq,
  midpoint,
  pointToBoundsDistance,
  pointToBoxDistance
} from "./chunk-SLG2OU3P.js";
import {
  distance,
  doSegmentsIntersect,
  doesLineIntersectLine,
  getSegmentIntersection,
  onSegment,
  orientation,
  pointToSegmentDistance
} from "./chunk-3453HRP7.js";
import {
  clamp,
  computeDistanceBetweenBoxes,
  findNearestPointsBetweenBoxSets,
  getBoundingBox
} from "./chunk-MHHTZHOJ.js";
import "./chunk-MTORG67J.js";
export {
  clamp,
  computeDistanceBetweenBoxes,
  distSq,
  distance,
  doSegmentsIntersect,
  doesLineIntersectLine,
  findNearestPointsBetweenBoxSets,
  getBoundingBox,
  getSegmentIntersection,
  getUnitVectorFromDirection,
  getUnitVectorFromPointAToB,
  grid,
  midpoint,
  onSegment,
  orientation,
  pointToBoundsDistance,
  pointToBoxDistance,
  pointToSegmentClosestPoint,
  pointToSegmentDistance,
  segmentToBoundsMinDistance,
  segmentToBoxMinDistance,
  segmentToCircleMinDistance,
  segmentToSegmentMinDistance
};
