// @sprucelabs/sprucebot-llm/build/chatWithImages.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const dotenv_1 = __importDefault(require("dotenv"));
const OpenAi_1 = __importDefault(require("./bots/adapters/OpenAi"));
const SprucebotLlmFactory_1 = __importDefault(require("./bots/SprucebotLlmFactory"));
dotenv_1.default.config();
void (async () => {
    console.clear();
    const image1 = fs_1.default.readFileSync('build/examples/images/image1.png');
    const image2 = fs_1.default.readFileSync('build/examples/images/image2.png');
    const base64Image1 = image1.toString('base64');
    const base64Image2 = image2.toString('base64');
    const adapter = OpenAi_1.default.Adapter(process.env.OPEN_AI_API_KEY);
    const bots = SprucebotLlmFactory_1.default.Factory(adapter);
    const skill = bots.Skill({
        weAreDoneWhen: 'you have described both images to me',
        yourJobIfYouChooseToAcceptItIs: 'to review the image i send you, invoke the getNextImage function and then describe that image too.',
        callbacks: {
            getNextImage: {
                cb: async () => {
                    return {
                        imageBase64: base64Image2,
                        imageDescription: 'A beautiful sunset over the mountains.',
                    };
                },
                useThisWhenever: 'you want to get the next image',
            },
        },
    });
    const bot = bots.Bot({
        skill,
        youAre: "a bot named Sprucebot that is in test mode. At the start of every conversation, you introduce yourself and announce that you are in test mode so I don't get confused! You are both hip and adorable. You say things like, 'Jeepers' and 'Golly' or even 'Jeezey peezy'!",
    });
    await bot.sendMessage({
        imageBase64: base64Image1,
        imageDescription: 'The first image',
    }, (message) => console.log('>', message));
})();
