// @sprucelabs/sprucebot-llm/build/chat.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_process_1 = require("node:process");
const readline = __importStar(require("node:readline/promises"));
const dotenv_1 = __importDefault(require("dotenv"));
const OpenAi_1 = __importDefault(require("./bots/adapters/OpenAi"));
const SprucebotLlmFactory_1 = __importDefault(require("./bots/SprucebotLlmFactory"));
const buildCallbackSkill_1 = __importDefault(require("./examples/buildCallbackSkill"));
const buildFileTransformerSkill_1 = __importDefault(require("./examples/buildFileTransformerSkill"));
const buildJokeSkill_1 = __importDefault(require("./examples/buildJokeSkill"));
const buildProfileSkill_1 = __importDefault(require("./examples/buildProfileSkill"));
const buildReceptionistSkill_1 = __importDefault(require("./examples/buildReceptionistSkill"));
dotenv_1.default.config();
const rl = readline.createInterface({ input: node_process_1.stdin, output: node_process_1.stdout });
void (async () => {
    console.clear();
    const adapter = OpenAi_1.default.Adapter(process.env.OPEN_AI_API_KEY);
    const bots = SprucebotLlmFactory_1.default.Factory(adapter);
    const skills = {
        jokes: (0, buildJokeSkill_1.default)(bots),
        profile: (0, buildProfileSkill_1.default)(bots),
        callbacks: (0, buildCallbackSkill_1.default)(bots),
        fileTransformer: (0, buildFileTransformerSkill_1.default)(bots),
        receptionist: (0, buildReceptionistSkill_1.default)(bots),
    };
    const bot = bots.Bot({
        skill: skills.callbacks,
        youAre: "a bot named Sprucebot that is in test mode. At the start of every conversation, you introduce yourself and announce that you are in test mode so I don't get confused! You are both hip and adorable. You say things like, 'Jeepers' and 'Golly' or even 'Jeezey peezy'!",
    });
    do {
        const input = await rl.question('Message > ');
        await bot.sendMessage(input, (message) => console.log('>', message));
    } while (!bot.getIsDone());
    console.log('Signing off...');
    rl.close();
})();
