// @sprucelabs/spruce-skill-utils/build/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SCHEMA_TYPES_FILENAME = exports.DEFAULT_SCHEMA_TYPES_FILE = exports.DEFAULT_BUILDER_FUNCTION = exports.DEFAULT_GLOBAL_SCHEMA_NAMESPACE = exports.SCHEMA_VERSION_FALLBACK = exports.ERROR_DIVIDER = exports.HEALTH_DIVIDER = exports.TEST_JEST_FAILED = exports.TEST_JEST_PASSED = exports.MERCURY_API_NAMESPACE = exports.CORE_NAMESPACE = exports.CORE_SCHEMA_VERSION = exports.LATEST_HANDLEBARS = exports.LATEST_TOKEN = exports.HASH_SPRUCE_BUILD_DIR = exports.HASH_SPRUCE_DIR = exports.HASH_SPRUCE_DIR_NAME = void 0;
exports.HASH_SPRUCE_DIR_NAME = '.spruce';
exports.HASH_SPRUCE_DIR = `src/.spruce`;
exports.HASH_SPRUCE_BUILD_DIR = `build/.spruce`;
exports.LATEST_TOKEN = '@latest';
exports.LATEST_HANDLEBARS = `{{${exports.LATEST_TOKEN}}}`;
exports.CORE_SCHEMA_VERSION = {
    intValue: 20200623,
    constValue: 'v2020_07_22',
    dirValue: 'v2020_07_22',
};
exports.CORE_NAMESPACE = 'Spruce';
exports.MERCURY_API_NAMESPACE = 'mercury';
exports.TEST_JEST_PASSED = 'passed';
exports.TEST_JEST_FAILED = 'failed';
exports.HEALTH_DIVIDER = '###### HEALTH REPORT ######';
exports.ERROR_DIVIDER = '###### ERROR DIVIDER ######';
exports.SCHEMA_VERSION_FALLBACK = '__latest';
exports.DEFAULT_GLOBAL_SCHEMA_NAMESPACE = 'SpruceSchemas';
exports.DEFAULT_BUILDER_FUNCTION = 'buildSchema';
exports.DEFAULT_SCHEMA_TYPES_FILE = '@sprucelabs/spruce-core-schemas';
exports.DEFAULT_SCHEMA_TYPES_FILENAME = 'schemas.types.ts';
