// @primer/react/lib/Truncate/Truncate.js
'use strict';

var React = require('react');
var clsx = require('clsx');
var BoxWithFallback = require('../internal/components/BoxWithFallback.js');
var Truncate_module = require('./Truncate.module.css.js');
var jsxRuntime = require('react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

const Truncate = /*#__PURE__*/React__default.default.forwardRef(function Truncate({
  as,
  children,
  className,
  title,
  inline,
  expandable,
  maxWidth = 125,
  style,
  sx,
  ...rest
}, ref) {
  return /*#__PURE__*/jsxRuntime.jsx(BoxWithFallback.BoxWithFallback, {
    ...rest,
    ref: ref,
    as: as,
    className: clsx.clsx(className, Truncate_module.Truncate),
    "data-expandable": expandable,
    "data-inline": inline,
    title: title,
    style: {
      ...style,
      [`--truncate-max-width`]: typeof maxWidth === 'number' ? `${maxWidth}px` : typeof maxWidth === 'string' ? maxWidth : undefined
    },
    sx: sx,
    children: children
  });
});
if (process.env.NODE_ENV !== "production") {
  Truncate.displayName = 'Truncate';
}

module.exports = Truncate;
