// @sprucelabs/spruce-core-schemas/build/esm/.spruce/schemas/spruce/v2020_07_22/feedItemTarget.schema.js
import { SchemaRegistry } from '@sprucelabs/schema';
const feedItemTargetSchema = {
    id: 'feedItemTarget',
    version: 'v2020_07_22',
    namespace: 'Spruce',
    name: '',
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-core-schemas',
    fields: {
        /** . */
        'personId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'casualName': {
            type: 'text',
            isRequired: true,
            options: undefined
        },
        /** . */
        'avatar': {
            type: 'image',
            options: { requiredSizes: ["*"], }
        },
        /** . */
        'skillId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'isMe': {
            type: 'boolean',
            isRequired: true,
            options: undefined
        },
        /** . */
        'isSprucebot': {
            type: 'boolean',
            isRequired: true,
            options: undefined
        },
    }
};
SchemaRegistry.getInstance().trackSchema(feedItemTargetSchema);
export default feedItemTargetSchema;
