// @sprucelabs/spruce-core-schemas/build/esm/.spruce/schemas/spruce/v2020_07_22/feedItem.schema.js
import { SchemaRegistry } from '@sprucelabs/schema';
import feedItemTargetSchema_v2020_07_22 from './feedItemTarget.schema.js';
import choiceSchema_v2020_07_22 from './choice.schema.js';
import linkSchema_v2020_07_22 from './link.schema.js';
const feedItemSchema = {
    id: 'feedItem',
    version: 'v2020_07_22',
    namespace: 'Spruce',
    name: '',
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-core-schemas',
    fields: {
        /** . */
        'id': {
            type: 'id',
            isRequired: true,
            options: undefined
        },
        /** . */
        'message': {
            type: 'text',
            isRequired: true,
            options: undefined
        },
        /** . */
        'dateCreated': {
            type: 'dateTime',
            isRequired: true,
            options: undefined
        },
        /** . */
        'note': {
            type: 'text',
            options: undefined
        },
        /** . */
        'source': {
            type: 'schema',
            isRequired: true,
            options: { schema: feedItemTargetSchema_v2020_07_22, }
        },
        /** . */
        'target': {
            type: 'schema',
            isRequired: true,
            options: { schema: feedItemTargetSchema_v2020_07_22, }
        },
        /** . */
        'choices': {
            type: 'schema',
            isArray: true,
            options: { schema: choiceSchema_v2020_07_22, }
        },
        /** . */
        'links': {
            type: 'schema',
            isArray: true,
            minArrayLength: 0,
            options: { schema: linkSchema_v2020_07_22, }
        },
        /** . */
        'context': {
            type: 'raw',
            options: { valueType: `Record<string, any>`, }
        },
    }
};
SchemaRegistry.getInstance().trackSchema(feedItemSchema);
export default feedItemSchema;
