// @sprucelabs/spruce-core-schemas/build/esm/.spruce/schemas/spruce/v2020_07_22/messageTarget.schema.js
import { SchemaRegistry } from '@sprucelabs/schema';
const messageTargetSchema = {
    id: 'messageTarget',
    version: 'v2020_07_22',
    namespace: 'Spruce',
    name: '',
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-core-schemas',
    fields: {
        /** . */
        'locationId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'personId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'organizationId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'skillId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'roleId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'phone': {
            type: 'phone',
            options: undefined
        },
    }
};
SchemaRegistry.getInstance().trackSchema(messageTargetSchema);
export default messageTargetSchema;
