// @sprucelabs/spruce-core-schemas/build/esm/.spruce/schemas/spruce/v2020_07_22/messageSource.schema.js
import { SchemaRegistry } from '@sprucelabs/schema';
const messageSourceSchema = {
    id: 'messageSource',
    version: 'v2020_07_22',
    namespace: 'Spruce',
    name: '',
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-core-schemas',
    fields: {
        /** . */
        'locationId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'personId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'organizationId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'skillId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'roleId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'isCore': {
            type: 'boolean',
            options: undefined
        },
    }
};
SchemaRegistry.getInstance().trackSchema(messageSourceSchema);
export default messageSourceSchema;
