// @sprucelabs/spruce-skill-utils/build/esm/interfaces/graphicsInterface.types.js
export var GraphicsTextEffect;
(function (GraphicsTextEffect) {
    GraphicsTextEffect["Reset"] = "reset";
    GraphicsTextEffect["Bold"] = "bold";
    GraphicsTextEffect["Dim"] = "dim";
    GraphicsTextEffect["Italic"] = "italic";
    GraphicsTextEffect["Underline"] = "underline";
    GraphicsTextEffect["Inverse"] = "inverse";
    GraphicsTextEffect["Hidden"] = "hidden";
    GraphicsTextEffect["Strikethrough"] = "strikethrough";
    GraphicsTextEffect["Visible"] = "visible";
    GraphicsTextEffect["Black"] = "black";
    GraphicsTextEffect["Red"] = "red";
    GraphicsTextEffect["Green"] = "green";
    GraphicsTextEffect["Yellow"] = "yellow";
    GraphicsTextEffect["Blue"] = "blue";
    GraphicsTextEffect["Magenta"] = "magenta";
    GraphicsTextEffect["Cyan"] = "cyan";
    GraphicsTextEffect["White"] = "white";
    GraphicsTextEffect["Gray"] = "gray";
    GraphicsTextEffect["Grey"] = "grey";
    GraphicsTextEffect["BlackBright"] = "blackBright";
    GraphicsTextEffect["RedBright"] = "redBright";
    GraphicsTextEffect["GreenBright"] = "greenBright";
    GraphicsTextEffect["YellowBright"] = "yellowBright";
    GraphicsTextEffect["BlueBright"] = "blueBright";
    GraphicsTextEffect["MagentaBright"] = "magentaBright";
    GraphicsTextEffect["CyanBright"] = "cyanBright";
    GraphicsTextEffect["WhiteBright"] = "whiteBright";
    GraphicsTextEffect["BgBlack"] = "bgBlack";
    GraphicsTextEffect["BgRed"] = "bgRed";
    GraphicsTextEffect["BgGreen"] = "bgGreen";
    GraphicsTextEffect["BgYellow"] = "bgYellow";
    GraphicsTextEffect["BgBlue"] = "bgBlue";
    GraphicsTextEffect["BgMagenta"] = "bgMagenta";
    GraphicsTextEffect["BgCyan"] = "bgCyan";
    GraphicsTextEffect["BgWhite"] = "bgWhite";
    GraphicsTextEffect["BgBlackBright"] = "bgBlackBright";
    GraphicsTextEffect["BgRedBright"] = "bgRedBright";
    GraphicsTextEffect["BgGreenBright"] = "bgGreenBright";
    GraphicsTextEffect["BgYellowBright"] = "bgYellowBright";
    GraphicsTextEffect["BgBlueBright"] = "bgBlueBright";
    GraphicsTextEffect["BgMagentaBright"] = "bgMagentaBright";
    GraphicsTextEffect["BgCyanBright"] = "bgCyanBright";
    GraphicsTextEffect["BgWhiteBright"] = "bgWhiteBright";
    GraphicsTextEffect["SpruceHeader"] = "tiny";
})(GraphicsTextEffect || (GraphicsTextEffect = {}));
