// @sprucelabs/mercury-types/build/esm/tests/StubClient.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
export default class StubClient {
    emitAndFlattenResponses(_eventName, _payload, _cb) {
        return __awaiter(this, void 0, void 0, function* () {
            return [];
        });
    }
    emit(_eventName, _payload, _cb) {
        return __awaiter(this, void 0, void 0, function* () {
            //@ts-ignore
            const results = {
                totalContracts: 1,
                totalResponses: 1,
                responses: [
                    {
                        responderRef: 'test',
                        payload: {
                            responsePayloadField: 'hello',
                        },
                    },
                ],
            };
            return results;
        });
    }
    on(_eventName, _cb) {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    off(_eventName) {
        return __awaiter(this, void 0, void 0, function* () {
            return 0;
        });
    }
}
