// @sprucelabs/spruce-skill-utils/build/esm/utilities/names.utility.js
import path from 'path';
import inflection from 'inflection';
import camelCase from 'lodash/camelCase.js';
import kebabCase from 'lodash/kebabCase.js';
import snakeCase from 'lodash/snakeCase.js';
import upperFirst from 'lodash/upperFirst.js';
const namesUtil = {
    /** Gets you a name good for using in an import statement based off a file path */
    toFileNameWithoutExtension(filePath) {
        return filePath
            .replace(path.dirname(filePath), '')
            .replace(path.extname(filePath), '')
            .replace('/', '');
    },
    /** First name => FirstName */
    toCamel(name) {
        return camelCase(name);
    },
    /** First name => FirstName */
    toPascal(name) {
        return upperFirst(this.toCamel(name));
    },
    /** First name => FIRST_NAME */
    toConst(name) {
        return snakeCase(name).toUpperCase();
    },
    /** First name => First names */
    toPlural(name) {
        return inflection.pluralize(name);
    },
    /** First name => First name */
    toSingular(name) {
        return inflection.singularize(name);
    },
    /** First name => first-name */
    toKebab(name) {
        return kebabCase(name);
    },
    /** First name => first_name */
    toSnake(name) {
        return snakeCase(name);
    },
};
export default namesUtil;
