// @primer/react/lib/ScrollableRegion/ScrollableRegion.js
'use strict';

var reactCompilerRuntime = require('react-compiler-runtime');
var React = require('react');
var useOverflow = require('../hooks/useOverflow.js');
var jsxRuntime = require('react/jsx-runtime');
var Box = require('../Box/Box.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

const defaultStyles = {
  // When setting overflow, we also set `position: relative` to avoid
  // `position: absolute` items breaking out of the container and causing
  // scrollbars on the page. This can occur with common classes like `sr-only`
  // and can cause difficult to track down layout issues
  position: 'relative',
  overflow: 'auto'
};
function ScrollableRegion(t0) {
  const $ = reactCompilerRuntime.c(13);
  let children;
  let label;
  let labelledby;
  let rest;
  if ($[0] !== t0) {
    ({
      "aria-label": label,
      "aria-labelledby": labelledby,
      children,
      ...rest
    } = t0);
    $[0] = t0;
    $[1] = children;
    $[2] = label;
    $[3] = labelledby;
    $[4] = rest;
  } else {
    children = $[1];
    label = $[2];
    labelledby = $[3];
    rest = $[4];
  }
  const ref = React__default.default.useRef(null);
  const hasOverflow = useOverflow.useOverflow(ref);
  let t1;
  if ($[5] !== hasOverflow || $[6] !== label || $[7] !== labelledby) {
    t1 = hasOverflow ? {
      "aria-label": label,
      "aria-labelledby": labelledby,
      role: "region",
      tabIndex: 0
    } : {};
    $[5] = hasOverflow;
    $[6] = label;
    $[7] = labelledby;
    $[8] = t1;
  } else {
    t1 = $[8];
  }
  const regionProps = t1;
  let t2;
  if ($[9] !== children || $[10] !== regionProps || $[11] !== rest) {
    t2 = /*#__PURE__*/jsxRuntime.jsx(Box, {
      ...rest,
      ...regionProps,
      ref: ref,
      sx: defaultStyles,
      children: children
    });
    $[9] = children;
    $[10] = regionProps;
    $[11] = rest;
    $[12] = t2;
  } else {
    t2 = $[12];
  }
  return t2;
}

exports.ScrollableRegion = ScrollableRegion;
