// @sprucelabs/schema/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapFieldErrorsToParameterErrors = exports.assertOptions = exports.fieldRegistrations = exports.SchemaRegistry = exports.SchemaEntityFactory = exports.FieldFactory = exports.expandValues = exports.flattenValues = exports.normalizeSchemaToIdWithVersion = exports.isIdWithVersion = exports.pickFields = exports.getFields = exports.formatPhoneNumber = exports.areSchemasTheSame = exports.isSchemaValid = exports.validateSchema = exports.dropPrivateFields = exports.dropFields = exports.makeFieldsOptional = exports.areSchemaValuesValid = exports.normalizeSchemaValues = exports.validateSchemaValues = exports.defaultSchemaValues = exports.registerFieldType = exports.buildSchema = exports.buildErrorSchema = exports.SchemaError = exports.KeyMapper = exports.selectAssertUtil = exports.cloneDeepPreservingInstances = exports.cloneDeep = exports.selectAssert = exports.validationErrorAssert = void 0;
__exportStar(require("./StaticSchemaEntityImpl"), exports);
const addressField_addon_1 = __importDefault(require("./addons/addressField.addon"));
const booleanField_addon_1 = __importDefault(require("./addons/booleanField.addon"));
const dateField_addon_1 = __importDefault(require("./addons/dateField.addon"));
const dateTimeField_addon_1 = __importDefault(require("./addons/dateTimeField.addon"));
const directoryField_addon_1 = __importDefault(require("./addons/directoryField.addon"));
const durationField_addon_1 = __importDefault(require("./addons/durationField.addon"));
const emailField_addon_1 = __importDefault(require("./addons/emailField.addon"));
const fileField_addon_1 = __importDefault(require("./addons/fileField.addon"));
const idField_addon_1 = __importDefault(require("./addons/idField.addon"));
const imageField_addon_1 = __importDefault(require("./addons/imageField.addon"));
const numberField_addon_1 = __importDefault(require("./addons/numberField.addon"));
const phoneField_addon_1 = __importDefault(require("./addons/phoneField.addon"));
const rawField_addon_1 = __importDefault(require("./addons/rawField.addon"));
const schemaField_addon_1 = __importDefault(require("./addons/schemaField.addon"));
const selectField_addon_1 = __importDefault(require("./addons/selectField.addon"));
const textField_addon_1 = __importDefault(require("./addons/textField.addon"));
const StaticSchemaEntityImpl_1 = __importDefault(require("./StaticSchemaEntityImpl"));
exports.default = StaticSchemaEntityImpl_1.default;
var validationErrorAssert_utility_1 = require("./tests/validationErrorAssert.utility");
Object.defineProperty(exports, "validationErrorAssert", { enumerable: true, get: function () { return __importDefault(validationErrorAssert_utility_1).default; } });
var selectAssert_utility_1 = require("./tests/selectAssert.utility");
Object.defineProperty(exports, "selectAssert", { enumerable: true, get: function () { return __importDefault(selectAssert_utility_1).default; } });
var cloneDeep_1 = require("./utilities/cloneDeep");
Object.defineProperty(exports, "cloneDeep", { enumerable: true, get: function () { return __importDefault(cloneDeep_1).default; } });
var cloneDeepPreservingInstances_1 = require("./utilities/cloneDeepPreservingInstances");
Object.defineProperty(exports, "cloneDeepPreservingInstances", { enumerable: true, get: function () { return __importDefault(cloneDeepPreservingInstances_1).default; } });
var selectAssert_deprecated_1 = require("./tests/selectAssert.deprecated");
Object.defineProperty(exports, "selectAssertUtil", { enumerable: true, get: function () { return __importDefault(selectAssert_deprecated_1).default; } });
var KeyMapper_1 = require("./utilities/KeyMapper");
Object.defineProperty(exports, "KeyMapper", { enumerable: true, get: function () { return __importDefault(KeyMapper_1).default; } });
__exportStar(require("./schemas.static.types"), exports);
__exportStar(require("./fields"), exports);
__exportStar(require("./errors/options.types"), exports);
var SpruceError_1 = require("./errors/SpruceError");
Object.defineProperty(exports, "SchemaError", { enumerable: true, get: function () { return __importDefault(SpruceError_1).default; } });
var buildErrorSchema_1 = require("./utilities/buildErrorSchema");
Object.defineProperty(exports, "buildErrorSchema", { enumerable: true, get: function () { return __importDefault(buildErrorSchema_1).default; } });
var buildSchema_1 = require("./utilities/buildSchema");
Object.defineProperty(exports, "buildSchema", { enumerable: true, get: function () { return __importDefault(buildSchema_1).default; } });
__exportStar(require("./utilities/optional.types"), exports);
__exportStar(require("./utilities/selectChoicesToHash"), exports);
__exportStar(require("./utilities/registerFieldType"), exports);
var registerFieldType_1 = require("./utilities/registerFieldType");
Object.defineProperty(exports, "registerFieldType", { enumerable: true, get: function () { return __importDefault(registerFieldType_1).default; } });
var defaultSchemaValues_1 = require("./utilities/defaultSchemaValues");
Object.defineProperty(exports, "defaultSchemaValues", { enumerable: true, get: function () { return __importDefault(defaultSchemaValues_1).default; } });
var validateSchemaValues_1 = require("./utilities/validateSchemaValues");
Object.defineProperty(exports, "validateSchemaValues", { enumerable: true, get: function () { return __importDefault(validateSchemaValues_1).default; } });
var normalizeSchemaValues_1 = require("./utilities/normalizeSchemaValues");
Object.defineProperty(exports, "normalizeSchemaValues", { enumerable: true, get: function () { return __importDefault(normalizeSchemaValues_1).default; } });
var areSchemaValuesValid_1 = require("./utilities/areSchemaValuesValid");
Object.defineProperty(exports, "areSchemaValuesValid", { enumerable: true, get: function () { return __importDefault(areSchemaValuesValid_1).default; } });
var makeFieldsOptional_1 = require("./utilities/makeFieldsOptional");
Object.defineProperty(exports, "makeFieldsOptional", { enumerable: true, get: function () { return __importDefault(makeFieldsOptional_1).default; } });
var dropFields_1 = require("./utilities/dropFields");
Object.defineProperty(exports, "dropFields", { enumerable: true, get: function () { return __importDefault(dropFields_1).default; } });
var dropPrivateFields_1 = require("./utilities/dropPrivateFields");
Object.defineProperty(exports, "dropPrivateFields", { enumerable: true, get: function () { return __importDefault(dropPrivateFields_1).default; } });
var validateSchema_1 = require("./utilities/validateSchema");
Object.defineProperty(exports, "validateSchema", { enumerable: true, get: function () { return __importDefault(validateSchema_1).default; } });
var isSchemaValid_1 = require("./utilities/isSchemaValid");
Object.defineProperty(exports, "isSchemaValid", { enumerable: true, get: function () { return __importDefault(isSchemaValid_1).default; } });
var areSchemasTheSame_1 = require("./utilities/areSchemasTheSame");
Object.defineProperty(exports, "areSchemasTheSame", { enumerable: true, get: function () { return __importDefault(areSchemasTheSame_1).default; } });
var formatPhoneNumber_1 = require("./utilities/formatPhoneNumber");
Object.defineProperty(exports, "formatPhoneNumber", { enumerable: true, get: function () { return __importDefault(formatPhoneNumber_1).default; } });
var getFields_1 = require("./utilities/getFields");
Object.defineProperty(exports, "getFields", { enumerable: true, get: function () { return __importDefault(getFields_1).default; } });
var pickFields_1 = require("./utilities/pickFields");
Object.defineProperty(exports, "pickFields", { enumerable: true, get: function () { return __importDefault(pickFields_1).default; } });
var isIdWithVersion_1 = require("./utilities/isIdWithVersion");
Object.defineProperty(exports, "isIdWithVersion", { enumerable: true, get: function () { return __importDefault(isIdWithVersion_1).default; } });
var normalizeSchemaToIdWithVersion_1 = require("./utilities/normalizeSchemaToIdWithVersion");
Object.defineProperty(exports, "normalizeSchemaToIdWithVersion", { enumerable: true, get: function () { return __importDefault(normalizeSchemaToIdWithVersion_1).default; } });
var flattenValues_1 = require("./utilities/flattenValues");
Object.defineProperty(exports, "flattenValues", { enumerable: true, get: function () { return __importDefault(flattenValues_1).default; } });
var expandValues_1 = require("./utilities/expandValues");
Object.defineProperty(exports, "expandValues", { enumerable: true, get: function () { return __importDefault(expandValues_1).default; } });
__exportStar(require("./utilities/formatPhoneNumber"), exports);
var FieldFactory_1 = require("./factories/FieldFactory");
Object.defineProperty(exports, "FieldFactory", { enumerable: true, get: function () { return __importDefault(FieldFactory_1).default; } });
var SchemaEntityFactory_1 = require("./factories/SchemaEntityFactory");
Object.defineProperty(exports, "SchemaEntityFactory", { enumerable: true, get: function () { return __importDefault(SchemaEntityFactory_1).default; } });
__exportStar(require("./types/template.types"), exports);
__exportStar(require("./types/utilities.types"), exports);
var SchemaRegistry_1 = require("./singletons/SchemaRegistry");
Object.defineProperty(exports, "SchemaRegistry", { enumerable: true, get: function () { return __importDefault(SchemaRegistry_1).default; } });
exports.fieldRegistrations = [
    addressField_addon_1.default,
    booleanField_addon_1.default,
    dateField_addon_1.default,
    dateTimeField_addon_1.default,
    directoryField_addon_1.default,
    durationField_addon_1.default,
    fileField_addon_1.default,
    idField_addon_1.default,
    numberField_addon_1.default,
    phoneField_addon_1.default,
    rawField_addon_1.default,
    schemaField_addon_1.default,
    selectField_addon_1.default,
    textField_addon_1.default,
    emailField_addon_1.default,
    imageField_addon_1.default,
];
var assertOptions_1 = require("./utilities/assertOptions");
Object.defineProperty(exports, "assertOptions", { enumerable: true, get: function () { return __importDefault(assertOptions_1).default; } });
var mapFieldErrorsToParameterErrors_1 = require("./utilities/mapFieldErrorsToParameterErrors");
Object.defineProperty(exports, "mapFieldErrorsToParameterErrors", { enumerable: true, get: function () { return __importDefault(mapFieldErrorsToParameterErrors_1).default; } });
