// @primer/react/lib/Header/Header.js
'use strict';

var reactCompilerRuntime = require('react-compiler-runtime');
var React = require('react');
var clsx = require('clsx');
var Header_module = require('./Header.module.css.js');
var BoxWithFallback = require('../internal/components/BoxWithFallback.js');
var jsxRuntime = require('react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

const Header = /*#__PURE__*/React__default.default.forwardRef(function Header(t0, forwardRef) {
  const $ = reactCompilerRuntime.c(13);
  let children;
  let className;
  let rest;
  let t1;
  if ($[0] !== t0) {
    ({
      children,
      className,
      as: t1,
      ...rest
    } = t0);
    $[0] = t0;
    $[1] = children;
    $[2] = className;
    $[3] = rest;
    $[4] = t1;
  } else {
    children = $[1];
    className = $[2];
    rest = $[3];
    t1 = $[4];
  }
  const as = t1 === undefined ? "header" : t1;
  let t2;
  if ($[5] !== className) {
    t2 = clsx.clsx(className, Header_module.Header);
    $[5] = className;
    $[6] = t2;
  } else {
    t2 = $[6];
  }
  let t3;
  if ($[7] !== as || $[8] !== children || $[9] !== forwardRef || $[10] !== rest || $[11] !== t2) {
    t3 = /*#__PURE__*/jsxRuntime.jsx(BoxWithFallback.BoxWithFallback, {
      as: as,
      ref: forwardRef,
      className: t2,
      ...rest,
      children: children
    });
    $[7] = as;
    $[8] = children;
    $[9] = forwardRef;
    $[10] = rest;
    $[11] = t2;
    $[12] = t3;
  } else {
    t3 = $[12];
  }
  return t3;
});
Header.displayName = 'Header';
const HeaderItem = /*#__PURE__*/React__default.default.forwardRef(function HeaderItem(t0, forwardRef) {
  const $ = reactCompilerRuntime.c(13);
  let children;
  let className;
  let full;
  let rest;
  if ($[0] !== t0) {
    ({
      children,
      className,
      full,
      ...rest
    } = t0);
    $[0] = t0;
    $[1] = children;
    $[2] = className;
    $[3] = full;
    $[4] = rest;
  } else {
    children = $[1];
    className = $[2];
    full = $[3];
    rest = $[4];
  }
  let t1;
  if ($[5] !== className) {
    t1 = clsx.clsx(className, Header_module.HeaderItem);
    $[5] = className;
    $[6] = t1;
  } else {
    t1 = $[6];
  }
  let t2;
  if ($[7] !== children || $[8] !== forwardRef || $[9] !== full || $[10] !== rest || $[11] !== t1) {
    t2 = /*#__PURE__*/jsxRuntime.jsx(BoxWithFallback.BoxWithFallback, {
      ref: forwardRef,
      className: t1,
      "data-full": full,
      ...rest,
      children: children
    });
    $[7] = children;
    $[8] = forwardRef;
    $[9] = full;
    $[10] = rest;
    $[11] = t1;
    $[12] = t2;
  } else {
    t2 = $[12];
  }
  return t2;
});
HeaderItem.displayName = 'Header.Item';
const HeaderLink = /*#__PURE__*/React__default.default.forwardRef(function HeaderLink(t0, forwardRef) {
  const $ = reactCompilerRuntime.c(13);
  let children;
  let className;
  let rest;
  let t1;
  if ($[0] !== t0) {
    ({
      children,
      className,
      as: t1,
      ...rest
    } = t0);
    $[0] = t0;
    $[1] = children;
    $[2] = className;
    $[3] = rest;
    $[4] = t1;
  } else {
    children = $[1];
    className = $[2];
    rest = $[3];
    t1 = $[4];
  }
  const as = t1 === undefined ? "a" : t1;
  let t2;
  if ($[5] !== className) {
    t2 = clsx.clsx(className, Header_module.HeaderLink);
    $[5] = className;
    $[6] = t2;
  } else {
    t2 = $[6];
  }
  let t3;
  if ($[7] !== as || $[8] !== children || $[9] !== forwardRef || $[10] !== rest || $[11] !== t2) {
    t3 = /*#__PURE__*/jsxRuntime.jsx(BoxWithFallback.BoxWithFallback, {
      as: as,
      ref: forwardRef,
      className: t2,
      ...rest,
      children: children
    });
    $[7] = as;
    $[8] = children;
    $[9] = forwardRef;
    $[10] = rest;
    $[11] = t2;
    $[12] = t3;
  } else {
    t3 = $[12];
  }
  return t3;
});
HeaderLink.displayName = 'Header.Link';
var Header$1 = Object.assign(Header, {
  Link: HeaderLink,
  Item: HeaderItem
});

module.exports = Header$1;
