// @tscircuit/simple-3d-svg/dist/index.js
// lib/index.ts
function colorToCss(c) {
  return typeof c === "string" ? c : `rgba(${c[0]},${c[1]},${c[2]},${c[3]})`;
}
function add(a, b) {
  return { x: a.x + b.x, y: a.y + b.y, z: a.z + b.z };
}
function sub(a, b) {
  return { x: a.x - b.x, y: a.y - b.y, z: a.z - b.z };
}
function dot(a, b) {
  return a.x * b.x + a.y * b.y + a.z * b.z;
}
function cross(a, b) {
  return {
    x: a.y * b.z - a.z * b.y,
    y: a.z * b.x - a.x * b.z,
    z: a.x * b.y - a.y * b.x
  };
}
function scale(v, k) {
  return { x: v.x * k, y: v.y * k, z: v.z * k };
}
function len(v) {
  return Math.sqrt(dot(v, v));
}
function norm(v) {
  const l = len(v) || 1;
  return scale(v, 1 / l);
}
function rotLocal(p, r = { x: 0, y: 0, z: 0 }) {
  let { x, y, z } = p;
  if (r.x) {
    const c = Math.cos(r.x);
    const s = Math.sin(r.x);
    const y2 = y * c - z * s;
    z = y * s + z * c;
    y = y2;
  }
  if (r.y) {
    const c = Math.cos(r.y);
    const s = Math.sin(r.y);
    const x2 = x * c + z * s;
    z = -x * s + z * c;
    x = x2;
  }
  if (r.z) {
    const c = Math.cos(r.z);
    const s = Math.sin(r.z);
    const x2 = x * c - y * s;
    y = x * s + y * c;
    x = x2;
  }
  return { x, y, z };
}
var W_DEF = 400;
var H_DEF = 400;
var FOCAL = 2;
function axes(cam) {
  const f = norm(sub(cam.lookAt, cam.position));
  const wUp = { x: 0, y: 1, z: 0 };
  let r = norm(cross(f, wUp));
  if (!len(r)) r = { x: 1, y: 0, z: 0 };
  const u = cross(r, f);
  return { r, u, f };
}
function toCam(p, cam) {
  const { r, u, f } = axes(cam);
  const d = sub(p, cam.position);
  return { x: dot(d, r), y: dot(d, u), z: dot(d, f) };
}
function proj(p, w, h, focal) {
  if (p.z <= 0) return null;
  const s = focal / p.z;
  return { x: p.x * s * w / 2, y: -p.y * s * h / 2, z: p.z };
}
var FACES = [
  [0, 1, 2, 3],
  [4, 5, 6, 7],
  [0, 1, 5, 4],
  [3, 2, 6, 7],
  [1, 2, 6, 5],
  [0, 3, 7, 4]
];
var TOP = [3, 2, 6, 7];
function verts(b) {
  const {
    size: { x: sx, y: sy, z: sz },
    center,
    rotation
  } = b;
  const offs = [
    { x: -sx / 2, y: -sy / 2, z: -sz / 2 },
    { x: sx / 2, y: -sy / 2, z: -sz / 2 },
    { x: sx / 2, y: sy / 2, z: -sz / 2 },
    { x: -sx / 2, y: sy / 2, z: -sz / 2 },
    { x: -sx / 2, y: -sy / 2, z: sz / 2 },
    { x: sx / 2, y: -sy / 2, z: sz / 2 },
    { x: sx / 2, y: sy / 2, z: sz / 2 },
    { x: -sx / 2, y: sy / 2, z: sz / 2 }
  ];
  return offs.map((o) => add(center, rotLocal(o, rotation)));
}
function renderScene(scene, opt = {}) {
  const W = opt.width ?? W_DEF;
  const H = opt.height ?? H_DEF;
  const focal = scene.camera.focalLength ?? FOCAL;
  const faces = [];
  const labels = [];
  for (const box of scene.boxes) {
    const vw = verts(box);
    const vc = vw.map((v) => toCam(v, scene.camera));
    const vp = vc.map((v) => proj(v, W, H, focal));
    for (const idx of FACES) {
      const p4 = [];
      let zMax = -Infinity;
      let behind = false;
      for (const i of idx) {
        const p = vp[i];
        if (!p) {
          behind = true;
          break;
        }
        p4.push(p);
        zMax = Math.max(zMax, vc[i].z);
      }
      if (behind) continue;
      const [a, b, c] = idx;
      const n = cross(sub(vc[b], vc[a]), sub(vc[c], vc[a]));
      if (n.z >= 0) continue;
      faces.push({
        pts: p4,
        depth: zMax,
        fill: colorToCss(box.color)
      });
    }
    if (box.topLabel) {
      const pts = TOP.map((i) => vp[i]);
      if (pts.every(Boolean)) {
        const p0 = pts[0];
        const p1 = pts[1];
        const p3 = pts[3];
        const u = sub(p1, p0);
        const v = sub(p3, p0);
        const lu = len(u);
        const lv = len(v);
        if (lu && lv) {
          const uN = scale(u, 1 / lu);
          const vN = scale(v, 1 / lv);
          const cx = pts.reduce((s, p) => s + p.x, 0) / 4;
          const cy = pts.reduce((s, p) => s + p.y, 0) / 4;
          const cz = Math.max(...TOP.map((i) => vc[i].z));
          const m = `matrix(${uN.x} ${uN.y} ${vN.x} ${vN.y} ${cx} ${cy})`;
          const fillCol = box.topLabelColor ?? [0, 0, 0, 1];
          labels.push({
            matrix: m,
            depth: cz,
            text: box.topLabel,
            fill: colorToCss(fillCol)
          });
        }
      }
    }
  }
  faces.sort((a, b) => b.depth - a.depth);
  labels.sort((a, b) => b.depth - a.depth);
  const out = [];
  out.push(
    `<svg xmlns="http://www.w3.org/2000/svg" width="${W}" height="${H}" viewBox="${-W / 2} ${-H / 2} ${W} ${H}">`
  );
  if (opt.backgroundColor) {
    out.push(
      `  <rect x="${-W / 2}" y="${-H / 2}" width="${W}" height="${H}" fill="${colorToCss(opt.backgroundColor)}" />
`
    );
  }
  out.push(
    '  <g stroke="#000" stroke-width="1" stroke-linecap="round" stroke-linejoin="round">\n'
  );
  for (const f of faces) {
    out.push(
      `    <polygon fill="${f.fill}" points="${f.pts.map((p) => `${p.x},${p.y}`).join(" ")}" />
`
    );
  }
  out.push("  </g>\n");
  out.push(
    '  <g font-family="sans-serif" font-size="14" text-anchor="middle" dominant-baseline="central">\n'
  );
  for (const l of labels) {
    out.push(
      `    <g transform="${l.matrix}"><text x="0" y="0" fill="${l.fill}">${l.text}</text></g>
`
    );
  }
  out.push("  </g>\n</svg>");
  return out.join("");
}
export {
  renderScene
};
