// @sprucelabs/spruce-core-schemas/build/.spruce/schemas/spruce/v2020_07_22/feedItem.schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const feedItemTarget_schema_1 = __importDefault(require("./feedItemTarget.schema"));
const choice_schema_1 = __importDefault(require("./choice.schema"));
const link_schema_1 = __importDefault(require("./link.schema"));
const feedItemSchema = {
    id: 'feedItem',
    version: 'v2020_07_22',
    namespace: 'Spruce',
    name: '',
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-core-schemas',
    fields: {
        /** . */
        'id': {
            type: 'id',
            isRequired: true,
            options: undefined
        },
        /** . */
        'message': {
            type: 'text',
            isRequired: true,
            options: undefined
        },
        /** . */
        'dateCreated': {
            type: 'dateTime',
            isRequired: true,
            options: undefined
        },
        /** . */
        'note': {
            type: 'text',
            options: undefined
        },
        /** . */
        'source': {
            type: 'schema',
            isRequired: true,
            options: { schema: feedItemTarget_schema_1.default, }
        },
        /** . */
        'target': {
            type: 'schema',
            isRequired: true,
            options: { schema: feedItemTarget_schema_1.default, }
        },
        /** . */
        'choices': {
            type: 'schema',
            isArray: true,
            options: { schema: choice_schema_1.default, }
        },
        /** . */
        'links': {
            type: 'schema',
            isArray: true,
            minArrayLength: 0,
            options: { schema: link_schema_1.default, }
        },
        /** . */
        'context': {
            type: 'raw',
            options: { valueType: `Record<string, any>`, }
        },
    }
};
schema_1.SchemaRegistry.getInstance().trackSchema(feedItemSchema);
exports.default = feedItemSchema;
