// @sprucelabs/spruce-core-schemas/build/.spruce/schemas/spruce/v2020_07_22/feedItemTarget.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const feedItemTargetSchema = {
    id: 'feedItemTarget',
    version: 'v2020_07_22',
    namespace: 'Spruce',
    name: '',
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-core-schemas',
    fields: {
        /** . */
        'personId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'casualName': {
            type: 'text',
            isRequired: true,
            options: undefined
        },
        /** . */
        'avatar': {
            type: 'image',
            options: { requiredSizes: ["*"], }
        },
        /** . */
        'skillId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'isMe': {
            type: 'boolean',
            isRequired: true,
            options: undefined
        },
        /** . */
        'isSprucebot': {
            type: 'boolean',
            isRequired: true,
            options: undefined
        },
    }
};
schema_1.SchemaRegistry.getInstance().trackSchema(feedItemTargetSchema);
exports.default = feedItemTargetSchema;
