// @sprucelabs/spruce-core-schemas/build/.spruce/schemas/spruce/v2020_07_22/messageTarget.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const messageTargetSchema = {
    id: 'messageTarget',
    version: 'v2020_07_22',
    namespace: 'Spruce',
    name: '',
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-core-schemas',
    fields: {
        /** . */
        'locationId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'personId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'organizationId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'skillId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'roleId': {
            type: 'id',
            options: undefined
        },
        /** . */
        'phone': {
            type: 'phone',
            options: undefined
        },
    }
};
schema_1.SchemaRegistry.getInstance().trackSchema(messageTargetSchema);
exports.default = messageTargetSchema;
