// @sprucelabs/spruce-core-schemas/build/.spruce/schemas/spruce/v2020_07_22/organization.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const organizationSchema = {
    id: 'organization',
    version: 'v2020_07_22',
    namespace: 'Spruce',
    name: 'Organization',
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-core-schemas',
    description: 'A company or team. Comprises of many people and locations.',
    fields: {
        /** Id. */
        'id': {
            label: 'Id',
            type: 'id',
            isRequired: true,
            options: undefined
        },
        /** Name. */
        'name': {
            label: 'Name',
            type: 'text',
            isRequired: true,
            options: undefined
        },
        /** Address. */
        'address': {
            label: 'Address',
            type: 'address',
            options: undefined
        },
        /** Slug. */
        'slug': {
            label: 'Slug',
            type: 'text',
            isRequired: true,
            options: undefined
        },
        /** Public. Is this organization viewable by guests? */
        'isPublic': {
            label: 'Public',
            type: 'boolean',
            hint: 'Is this organization viewable by guests?',
            defaultValue: false,
            options: undefined
        },
        /** . */
        'dateCreated': {
            type: 'dateTime',
            isRequired: true,
            options: undefined
        },
        /** . */
        'dateDeleted': {
            type: 'dateTime',
            options: undefined
        },
    }
};
schema_1.SchemaRegistry.getInstance().trackSchema(organizationSchema);
exports.default = organizationSchema;
