// @sprucelabs/spruce-core-schemas/build/.spruce/schemas/spruce/v2020_07_22/personLocation.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const personLocationSchema = {
    id: 'personLocation',
    version: 'v2020_07_22',
    namespace: 'Spruce',
    name: 'Person <-> location relationship',
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-core-schemas',
    fields: {
        /** Id. */
        'id': {
            label: 'Id',
            type: 'id',
            isRequired: true,
            options: undefined
        },
        /** Location id. */
        'locationId': {
            label: 'Location id',
            type: 'id',
            isRequired: true,
            options: undefined
        },
        /** Person id. */
        'personId': {
            label: 'Person id',
            type: 'id',
            isRequired: true,
            options: undefined
        },
        /** Date created. */
        'dateCreated': {
            label: 'Date created',
            type: 'dateTime',
            isRequired: true,
            options: undefined
        },
        /** Date deleted. */
        'dateDeleted': {
            label: 'Date deleted',
            type: 'dateTime',
            options: undefined
        },
    }
};
schema_1.SchemaRegistry.getInstance().trackSchema(personLocationSchema);
exports.default = personLocationSchema;
