// @sprucelabs/spruce-core-schemas/build/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BASE_ROLE_CHOICES_WITH_META = exports.BASE_ROLE_CHOICES = exports.META_BASE_ROLE_CHOICES = exports.GUEST_BASE_ROLE_CHOICES = exports.EMPLOYED_BASE_ROLE_CHOICES = exports.BASE_ROLES_WITH_META = exports.BASE_ROLES = exports.META_BASE_ROLES = exports.GUEST_BASE_ROLES = exports.EMPLOYED_BASE_ROLES = void 0;
exports.EMPLOYED_BASE_ROLES = [
    { name: 'Owner', slug: 'owner' },
    { name: 'Group manager', slug: 'groupManager' },
    { name: 'Manager', slug: 'manager' },
    { name: 'Teammate', slug: 'teammate' },
];
exports.GUEST_BASE_ROLES = [{ name: 'Guest', slug: 'guest' }];
exports.META_BASE_ROLES = [
    { name: 'Anonymous', slug: 'anonymous' },
    { name: 'Logged in', slug: 'loggedIn' },
];
exports.BASE_ROLES = [...exports.EMPLOYED_BASE_ROLES, ...exports.GUEST_BASE_ROLES];
exports.BASE_ROLES_WITH_META = [
    ...exports.EMPLOYED_BASE_ROLES,
    ...exports.GUEST_BASE_ROLES,
    ...exports.META_BASE_ROLES,
];
exports.EMPLOYED_BASE_ROLE_CHOICES = [
    { label: 'Owner', value: 'owner' },
    { label: 'Group manager', value: 'groupManager' },
    { label: 'Manager', value: 'manager' },
    { label: 'Teammate', value: 'teammate' },
];
exports.GUEST_BASE_ROLE_CHOICES = [
    { label: 'Guest', value: 'guest' },
];
exports.META_BASE_ROLE_CHOICES = [
    { label: 'Anonymous', value: 'anonymous' },
    { label: 'Logged in', value: 'loggedIn' },
];
exports.BASE_ROLE_CHOICES = [
    ...exports.EMPLOYED_BASE_ROLE_CHOICES,
    ...exports.GUEST_BASE_ROLE_CHOICES,
];
exports.BASE_ROLE_CHOICES_WITH_META = [
    ...exports.EMPLOYED_BASE_ROLE_CHOICES,
    ...exports.META_BASE_ROLE_CHOICES,
    ...exports.GUEST_BASE_ROLE_CHOICES,
];
