// @sprucelabs/spruce-core-schemas/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.timezoneChoices = exports.choiceSchema = exports.feedItemSchema = exports.linkSchema = exports.feedItemTargetSchema = exports.feedSchema = exports.sendMessageSchema = exports.messageTargetSchema = exports.messageSourceSchema = exports.messageSchema = exports.roleSchema = exports.personLocationSchema = exports.personOrganizationSchema = exports.organizationSchema = exports.locationSchema = exports.skillCreatorSchema = exports.skillSchema = exports.personSchema = void 0;
__exportStar(require("./.spruce/schemas/core.schemas.types"), exports);
var person_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/person.schema");
Object.defineProperty(exports, "personSchema", { enumerable: true, get: function () { return __importDefault(person_schema_1).default; } });
var skill_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/skill.schema");
Object.defineProperty(exports, "skillSchema", { enumerable: true, get: function () { return __importDefault(skill_schema_1).default; } });
var skillCreator_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/skillCreator.schema");
Object.defineProperty(exports, "skillCreatorSchema", { enumerable: true, get: function () { return __importDefault(skillCreator_schema_1).default; } });
var location_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/location.schema");
Object.defineProperty(exports, "locationSchema", { enumerable: true, get: function () { return __importDefault(location_schema_1).default; } });
var organization_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/organization.schema");
Object.defineProperty(exports, "organizationSchema", { enumerable: true, get: function () { return __importDefault(organization_schema_1).default; } });
var personOrganization_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/personOrganization.schema");
Object.defineProperty(exports, "personOrganizationSchema", { enumerable: true, get: function () { return __importDefault(personOrganization_schema_1).default; } });
var personLocation_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/personLocation.schema");
Object.defineProperty(exports, "personLocationSchema", { enumerable: true, get: function () { return __importDefault(personLocation_schema_1).default; } });
var role_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/role.schema");
Object.defineProperty(exports, "roleSchema", { enumerable: true, get: function () { return __importDefault(role_schema_1).default; } });
var message_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/message.schema");
Object.defineProperty(exports, "messageSchema", { enumerable: true, get: function () { return __importDefault(message_schema_1).default; } });
var messageSource_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/messageSource.schema");
Object.defineProperty(exports, "messageSourceSchema", { enumerable: true, get: function () { return __importDefault(messageSource_schema_1).default; } });
var messageTarget_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/messageTarget.schema");
Object.defineProperty(exports, "messageTargetSchema", { enumerable: true, get: function () { return __importDefault(messageTarget_schema_1).default; } });
var sendMessage_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/sendMessage.schema");
Object.defineProperty(exports, "sendMessageSchema", { enumerable: true, get: function () { return __importDefault(sendMessage_schema_1).default; } });
var feed_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/feed.schema");
Object.defineProperty(exports, "feedSchema", { enumerable: true, get: function () { return __importDefault(feed_schema_1).default; } });
var feedItemTarget_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/feedItemTarget.schema");
Object.defineProperty(exports, "feedItemTargetSchema", { enumerable: true, get: function () { return __importDefault(feedItemTarget_schema_1).default; } });
var link_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/link.schema");
Object.defineProperty(exports, "linkSchema", { enumerable: true, get: function () { return __importDefault(link_schema_1).default; } });
var feedItem_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/feedItem.schema");
Object.defineProperty(exports, "feedItemSchema", { enumerable: true, get: function () { return __importDefault(feedItem_schema_1).default; } });
var choice_schema_1 = require("./.spruce/schemas/spruce/v2020_07_22/choice.schema");
Object.defineProperty(exports, "choiceSchema", { enumerable: true, get: function () { return __importDefault(choice_schema_1).default; } });
var timezoneChoices_1 = require("./timezoneChoices");
Object.defineProperty(exports, "timezoneChoices", { enumerable: true, get: function () { return timezoneChoices_1.timezoneChoices; } });
__exportStar(require("./constants"), exports);
__exportStar(require("./core.types"), exports);
