// @sprucelabs/spruce-core-schemas/build/esm/schemas/v2020_07_22/message.builder.js
import { buildSchema } from '@sprucelabs/schema';
import choiceBuilder from './choice.builder.js';
import linkBuilder from './link.builder.js';
const eventTargetFields = {
    locationId: {
        type: 'id',
    },
    personId: {
        type: 'id',
    },
    organizationId: {
        type: 'id',
    },
    skillId: {
        type: 'id',
    },
    roleId: {
        type: 'id',
    },
};
export const messageTargetSchema = buildSchema({
    id: 'messageTarget',
    fields: Object.assign(Object.assign({}, eventTargetFields), { phone: {
            type: 'phone',
        } }),
});
export const messageSourceSchema = buildSchema({
    id: 'messageSource',
    fields: Object.assign(Object.assign({}, eventTargetFields), { isCore: {
            type: 'boolean',
        } }),
});
export default buildSchema({
    id: 'message',
    description: 'A discrete communication between two humans or a human and a machine.',
    fields: {
        id: {
            type: 'id',
            isRequired: true,
        },
        trackingId: {
            type: 'id',
            hint: 'An arbitrary id that can be used to track this message when being sent and not yet assigned an Id.',
        },
        dateCreated: {
            type: 'number',
            isRequired: true,
        },
        dateSent: {
            type: 'number',
        },
        target: {
            type: 'schema',
            isRequired: true,
            options: {
                schema: messageTargetSchema,
            },
        },
        source: {
            type: 'schema',
            isRequired: true,
            options: {
                schema: messageSourceSchema,
            },
        },
        errors: {
            type: 'text',
            isArray: true,
            isPrivate: true,
        },
        classification: {
            type: 'select',
            isRequired: true,
            options: {
                choices: [
                    {
                        value: 'auth',
                        label: 'Auth',
                    },
                    {
                        value: 'transactional',
                        label: 'transactional',
                    },
                    {
                        value: 'promotional',
                        label: 'Promotional',
                    },
                    {
                        value: 'incoming',
                        label: 'incoming',
                    },
                ],
            },
        },
        status: {
            type: 'select',
            defaultValue: 'pending',
            isPrivate: true,
            options: {
                choices: [
                    {
                        value: 'pending',
                        label: 'Pending',
                    },
                    {
                        value: 'processing',
                        label: 'Processing',
                    },
                    {
                        value: 'sent',
                        label: 'Sent',
                    },
                    {
                        value: 'failed',
                        label: 'Failed',
                    },
                    {
                        value: 'ignored',
                        label: 'Ignored',
                    },
                ],
            },
        },
        body: {
            type: 'text',
            isRequired: true,
        },
        context: {
            type: 'raw',
            options: {
                valueType: 'Record<string, any>',
            },
            isPrivate: true,
        },
        topicId: {
            type: 'id',
        },
        links: {
            type: 'schema',
            isArray: true,
            minArrayLength: 0,
            options: {
                schema: linkBuilder,
            },
        },
        choices: {
            type: 'schema',
            isArray: true,
            options: {
                schema: choiceBuilder,
            },
        },
    },
});
