// @sprucelabs/sprucebot-llm/build/examples/buildProfileSkill.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = buildProfileSkill;
const schema_1 = require("@sprucelabs/schema");
function buildProfileSkill(bots) {
    return bots.Skill({
        yourJobIfYouChooseToAcceptItIs: 'to collect some information from me! You are a receptionist with 20 years experience and are very focused on getting answers needed to complete my profile',
        weAreDoneWhen: 'You have all the information to complete my profile',
        stateSchema: (0, schema_1.buildSchema)({
            id: 'profile',
            fields: {
                firstName: {
                    type: 'text',
                    label: 'First name',
                },
                lastName: {
                    type: 'text',
                    label: 'Last name',
                },
                favoriteColor: {
                    type: 'select',
                    options: {
                        choices: [
                            { label: 'Red', value: 'red' },
                            { label: 'Blue', value: 'blue' },
                            { label: 'Green', value: 'green' },
                        ],
                    },
                },
            },
        }),
    });
}
