// @sprucelabs/sprucebot-llm/build/bots/adapters/SpyOpenAiApi.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const openai_1 = __importDefault(require("openai"));
class SpyOpenAiApi extends openai_1.default {
    constructor(config) {
        super(config);
        //@ts-ignore
        this.chat = {
            completions: {
                create: this.createCompletion.bind(this),
            },
        };
        SpyOpenAiApi.config = config;
    }
    async createCompletion(options) {
        SpyOpenAiApi.lastSentCompletion = options;
        const choices = [];
        if (SpyOpenAiApi.responseMessage) {
            choices.push({
                finish_reason: 'stop',
                index: 0,
                logprobs: null,
                message: {
                    content: SpyOpenAiApi.responseMessage,
                    role: 'assistant',
                    refusal: null,
                },
            });
        }
        return {
            id: 'cmpl-1',
            model: 'text-davinci-003',
            created: 0,
            choices,
            object: 'chat.completion',
        };
    }
}
SpyOpenAiApi.responseMessage = 'hello!';
exports.default = SpyOpenAiApi;
