// @sprucelabs/schema/build/esm/.spruce/schemas/fields/fieldClassMap.js
import AddressField from '../../../fields/AddressField.js';
import BooleanField from '../../../fields/BooleanField.js';
import DateField from '../../../fields/DateField.js';
import DateTimeField from '../../../fields/DateTimeField.js';
import DirectoryField from '../../../fields/DirectoryField.js';
import DurationField from '../../../fields/DurationField.js';
import EmailField from '../../../fields/EmailField.js';
import FileField from '../../../fields/FileField.js';
import IdField from '../../../fields/IdField.js';
import ImageField from '../../../fields/ImageField.js';
import NumberField from '../../../fields/NumberField.js';
import PhoneField from '../../../fields/PhoneField.js';
import RawField from '../../../fields/RawField.js';
import SchemaField from '../../../fields/SchemaField.js';
import SelectField from '../../../fields/SelectField.js';
import TextField from '../../../fields/TextField.js';
const fieldClassMap = {
    boolean: BooleanField,
    select: SelectField,
    duration: DurationField,
    id: IdField,
    address: AddressField,
    phone: PhoneField,
    schema: SchemaField,
    raw: RawField,
    number: NumberField,
    dateTime: DateTimeField,
    text: TextField,
    file: FileField,
    date: DateField,
    directory: DirectoryField,
    email: EmailField,
    image: ImageField,
};
export default fieldClassMap;
