// @primer/react/lib/InlineMessage/InlineMessage.js
'use strict';

var reactCompilerRuntime = require('react-compiler-runtime');
var octiconsReact = require('@primer/octicons-react');
var clsx = require('clsx');
var InlineMessage_module = require('./InlineMessage.module.css.js');
var BoxWithFallback = require('../internal/components/BoxWithFallback.js');
var jsxRuntime = require('react/jsx-runtime');

const variantToIcon = variant => {
  const icons = {
    warning: /*#__PURE__*/jsxRuntime.jsx(octiconsReact.AlertIcon, {
      className: InlineMessage_module.InlineMessageIcon
    }),
    critical: /*#__PURE__*/jsxRuntime.jsx(octiconsReact.AlertIcon, {
      className: InlineMessage_module.InlineMessageIcon
    }),
    success: /*#__PURE__*/jsxRuntime.jsx(octiconsReact.CheckCircleIcon, {
      className: InlineMessage_module.InlineMessageIcon
    }),
    unavailable: /*#__PURE__*/jsxRuntime.jsx(octiconsReact.AlertIcon, {
      className: InlineMessage_module.InlineMessageIcon
    })
  };
  return icons[variant];
};
const variantToSmallIcon = variant => {
  const icons = {
    warning: /*#__PURE__*/jsxRuntime.jsx(octiconsReact.AlertFillIcon, {
      className: InlineMessage_module.InlineMessageIcon,
      size: 12
    }),
    critical: /*#__PURE__*/jsxRuntime.jsx(octiconsReact.AlertFillIcon, {
      className: InlineMessage_module.InlineMessageIcon,
      size: 12
    }),
    success: /*#__PURE__*/jsxRuntime.jsx(octiconsReact.CheckCircleFillIcon, {
      className: InlineMessage_module.InlineMessageIcon,
      size: 12
    }),
    unavailable: /*#__PURE__*/jsxRuntime.jsx(octiconsReact.AlertFillIcon, {
      className: InlineMessage_module.InlineMessageIcon,
      size: 12
    })
  };
  return icons[variant];
};
function InlineMessage(t0) {
  const $ = reactCompilerRuntime.c(18);
  let children;
  let className;
  let rest;
  let t1;
  let variant;
  if ($[0] !== t0) {
    ({
      children,
      className,
      size: t1,
      variant,
      ...rest
    } = t0);
    $[0] = t0;
    $[1] = children;
    $[2] = className;
    $[3] = rest;
    $[4] = t1;
    $[5] = variant;
  } else {
    children = $[1];
    className = $[2];
    rest = $[3];
    t1 = $[4];
    variant = $[5];
  }
  const size = t1 === undefined ? "medium" : t1;
  let t2;
  if ($[6] !== size || $[7] !== variant) {
    t2 = size === "small" ? variantToSmallIcon(variant) : variantToIcon(variant);
    $[6] = size;
    $[7] = variant;
    $[8] = t2;
  } else {
    t2 = $[8];
  }
  const icon = t2;
  let t3;
  if ($[9] !== className) {
    t3 = clsx.clsx(className, InlineMessage_module.InlineMessage);
    $[9] = className;
    $[10] = t3;
  } else {
    t3 = $[10];
  }
  let t4;
  if ($[11] !== children || $[12] !== icon || $[13] !== rest || $[14] !== size || $[15] !== t3 || $[16] !== variant) {
    t4 = /*#__PURE__*/jsxRuntime.jsxs(BoxWithFallback.BoxWithFallback, {
      className: t3,
      ...rest,
      "data-size": size,
      "data-variant": variant,
      children: [icon, children]
    });
    $[11] = children;
    $[12] = icon;
    $[13] = rest;
    $[14] = size;
    $[15] = t3;
    $[16] = variant;
    $[17] = t4;
  } else {
    t4 = $[17];
  }
  return t4;
}

exports.InlineMessage = InlineMessage;
