// @sprucelabs/spruce-event-utils/build/schemas/v2021_09_13/eventPagingRequest.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@sprucelabs/schema");
const sortSchema = (0, schema_1.buildSchema)({
    id: 'eventPagingSort',
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-event-utils',
    importsWhenRemote: ["import '@sprucelabs/spruce-event-utils'"],
    fields: {
        field: {
            type: 'id',
            isRequired: true,
        },
        direction: {
            type: 'select',
            options: {
                choices: [
                    {
                        value: 'asc',
                        label: 'Ascending',
                    },
                    {
                        value: 'desc',
                        label: 'Descending',
                    },
                ],
            },
        },
    },
});
exports.default = (0, schema_1.buildSchema)({
    id: 'eventPagingRequest',
    name: 'Event paging request',
    moduleToImportFromWhenRemote: '@sprucelabs/spruce-event-utils',
    importsWhenRemote: ["import '@sprucelabs/spruce-event-utils'"],
    fields: {
        pageSize: {
            type: 'number',
        },
        next: {
            type: 'id',
        },
        previous: {
            type: 'id',
        },
        sort: {
            type: 'schema',
            isArray: true,
            options: {
                schema: sortSchema,
            },
        },
    },
});
