// @sprucelabs/spruce-event-utils/build/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.REMOTES = exports.REMOTE_PROD = exports.REMOTE_POLISH = exports.REMOTE_SANDBOX = exports.REMOTE_HEARTWOOD_TEST = exports.REMOTE_DEMO = exports.REMOTE_DEV = exports.REMOTE_LOCAL = exports.EVENT_VERSION_DIVIDER = void 0;
exports.EVENT_VERSION_DIVIDER = '::';
exports.REMOTE_LOCAL = 'http://127.0.0.1:8081';
exports.REMOTE_DEV = 'https://developer.mercury.spruce.ai';
exports.REMOTE_DEMO = 'https://demo.mercury.spruce.ai';
exports.REMOTE_HEARTWOOD_TEST = 'https://heartwood.mercury.spruce.ai';
exports.REMOTE_SANDBOX = 'https://sandbox.mercury.spruce.ai';
exports.REMOTE_POLISH = 'https://polish.mercury.spruce.ai';
exports.REMOTE_PROD = 'https://mercury.spruce.ai';
exports.REMOTES = {
    local: exports.REMOTE_LOCAL,
    developer: exports.REMOTE_DEV,
    demo: exports.REMOTE_DEMO,
    sandbox: exports.REMOTE_SANDBOX,
    polish: exports.REMOTE_POLISH,
    production: exports.REMOTE_PROD,
    heartwood_test: exports.REMOTE_HEARTWOOD_TEST,
};
