// @primer/react/lib/Flash/Flash.js
'use strict';

var reactCompilerRuntime = require('react-compiler-runtime');
var clsx = require('clsx');
var React = require('react');
var BoxWithFallback = require('../internal/components/BoxWithFallback.js');
var Flash_module = require('./Flash.module.css.js');
var jsxRuntime = require('react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

const Flash = /*#__PURE__*/React__default.default.forwardRef(function Flash(t0, ref) {
  const $ = reactCompilerRuntime.c(17);
  let as;
  let className;
  let full;
  let rest;
  let sx;
  let t1;
  if ($[0] !== t0) {
    ({
      as,
      className,
      variant: t1,
      full,
      sx,
      ...rest
    } = t0);
    $[0] = t0;
    $[1] = as;
    $[2] = className;
    $[3] = full;
    $[4] = rest;
    $[5] = sx;
    $[6] = t1;
  } else {
    as = $[1];
    className = $[2];
    full = $[3];
    rest = $[4];
    sx = $[5];
    t1 = $[6];
  }
  const variant = t1 === undefined ? "default" : t1;
  let t2;
  if ($[7] !== className) {
    t2 = clsx.clsx(Flash_module.Flash, className);
    $[7] = className;
    $[8] = t2;
  } else {
    t2 = $[8];
  }
  const t3 = full ? "" : undefined;
  let t4;
  if ($[9] !== as || $[10] !== ref || $[11] !== rest || $[12] !== sx || $[13] !== t2 || $[14] !== t3 || $[15] !== variant) {
    t4 = /*#__PURE__*/jsxRuntime.jsx(BoxWithFallback.BoxWithFallback, {
      ...rest,
      ref: ref,
      as: as,
      className: t2,
      "data-full": t3,
      "data-variant": variant,
      sx: sx
    });
    $[9] = as;
    $[10] = ref;
    $[11] = rest;
    $[12] = sx;
    $[13] = t2;
    $[14] = t3;
    $[15] = variant;
    $[16] = t4;
  } else {
    t4 = $[16];
  }
  return t4;
});
if (process.env.NODE_ENV !== "production") {
  Flash.displayName = 'Flash';
}

module.exports = Flash;
