// @primer/react/lib/experimental/UnderlinePanels/UnderlinePanels.js
'use strict';

var reactCompilerRuntime = require('react-compiler-runtime');
var React = require('react');
var tabContainerElement = require('../../node_modules/@github/tab-container-element/dist/tab-container-element.js');
require('../../node_modules/@github/tab-container-element/dist/tab-container-element-define.js');
var createComponent = require('../../utils/create-component.js');
var UnderlineTabbedInterface = require('../../internal/components/UnderlineTabbedInterface.js');
var invariant = require('../../utils/invariant.js');
var useResizeObserver = require('../../hooks/useResizeObserver.js');
var useIsomorphicLayoutEffect = require('../../utils/useIsomorphicLayoutEffect.js');
var UnderlinePanels_module = require('./UnderlinePanels.module.css.js');
var clsx = require('clsx');
var BoxWithFallback = require('../../internal/components/BoxWithFallback.js');
var jsxRuntime = require('react/jsx-runtime');
var useId = require('../../hooks/useId.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

const TabContainerComponent = createComponent.createComponent(tabContainerElement.TabContainerElement, 'tab-container');
const UnderlinePanels = ({
  'aria-label': ariaLabel,
  'aria-labelledby': ariaLabelledBy,
  children,
  loadingCounters,
  className,
  ...props
}) => {
  const [iconsVisible, setIconsVisible] = React.useState(true);
  const wrapperRef = React.useRef(null);
  const listRef = React.useRef(null);
  // We need to always call useId() because React Hooks must be
  // called in the exact same order in every component render
  const parentId = useId.useId(props.id);
  const [tabs, setTabs] = React.useState([]);
  const [tabPanels, setTabPanels] = React.useState([]);

  // Make sure we have fresh prop data whenever the tabs or panels are updated (keep aria-selected current)
  React.useEffect(() => {
    // Loop through the chidren, if it's a tab, then add id="{id}-tab-{index}"
    // If it's a panel, then add aria-labelledby="{id}-tab-{index}"
    let tabIndex = 0;
    let panelIndex = 0;
    const childrenWithProps = React.Children.map(children, child => {
      if (/*#__PURE__*/React.isValidElement(child) && child.type === Tab) {
        return /*#__PURE__*/React.cloneElement(child, {
          id: `${parentId}-tab-${tabIndex++}`,
          loadingCounters,
          iconsVisible
        });
      }
      if (/*#__PURE__*/React.isValidElement(child) && child.type === Panel) {
        return /*#__PURE__*/React.cloneElement(child, {
          'aria-labelledby': `${parentId}-tab-${panelIndex++}`
        });
      }
      return child;
    });
    const newTabs = React.Children.toArray(childrenWithProps).filter(child_0 => {
      return /*#__PURE__*/React.isValidElement(child_0) && child_0.type === Tab;
    });
    const newTabPanels = React.Children.toArray(childrenWithProps).filter(child_1 => /*#__PURE__*/React.isValidElement(child_1) && child_1.type === Panel);
    setTabs(newTabs);
    setTabPanels(newTabPanels);
  }, [children, parentId, loadingCounters, iconsVisible]);
  const tabsHaveIcons = tabs.some(tab => /*#__PURE__*/React__default.default.isValidElement(tab) && tab.props.icon);

  // this is a workaround to get the list's width on the first render
  const [listWidth, setListWidth] = React.useState(0);
  useIsomorphicLayoutEffect(() => {
    var _listRef$current$getB, _listRef$current;
    if (!tabsHaveIcons) {
      return;
    }
    setListWidth((_listRef$current$getB = (_listRef$current = listRef.current) === null || _listRef$current === void 0 ? void 0 : _listRef$current.getBoundingClientRect().width) !== null && _listRef$current$getB !== void 0 ? _listRef$current$getB : 0);
  }, [tabsHaveIcons]);

  // when the wrapper resizes, check if the icons should be visible
  // by comparing the wrapper width to the list width
  useResizeObserver.useResizeObserver(resizeObserverEntries => {
    if (!tabsHaveIcons) {
      return;
    }
    const wrapperWidth = resizeObserverEntries[0].contentRect.width;
    setIconsVisible(wrapperWidth > listWidth);
  }, wrapperRef, []);
  if (process.env.NODE_ENV !== "production") {
    const selectedTabs = tabs.filter(tab_0 => {
      const ariaSelected = /*#__PURE__*/React__default.default.isValidElement(tab_0) && tab_0.props['aria-selected'];
      return ariaSelected === true || ariaSelected === 'true';
    });
    !(selectedTabs.length <= 1) ? process.env.NODE_ENV !== "production" ? invariant.invariant(false, 'Only one tab can be selected at a time.') : invariant.invariant(false) : void 0;
    !(tabs.length === tabPanels.length) ? process.env.NODE_ENV !== "production" ? invariant.invariant(false, `The number of tabs and panels must be equal. Counted ${tabs.length} tabs and ${tabPanels.length} panels.`) : invariant.invariant(false) : void 0;
  }
  return /*#__PURE__*/jsxRuntime.jsxs(TabContainerComponent, {
    children: [/*#__PURE__*/jsxRuntime.jsx(UnderlineTabbedInterface.UnderlineWrapper, {
      ref: wrapperRef,
      slot: "tablist-wrapper",
      "data-icons-visible": iconsVisible,
      className: clsx.clsx(className, UnderlinePanels_module.StyledUnderlineWrapper),
      ...props,
      children: /*#__PURE__*/jsxRuntime.jsx(UnderlineTabbedInterface.UnderlineItemList, {
        ref: listRef,
        "aria-label": ariaLabel,
        "aria-labelledby": ariaLabelledBy,
        role: "tablist",
        children: tabs
      })
    }), tabPanels]
  });
};
UnderlinePanels.displayName = "UnderlinePanels";
const Tab = t0 => {
  const $ = reactCompilerRuntime.c(14);
  let ariaSelected;
  let onSelect;
  let props;
  if ($[0] !== t0) {
    ({
      "aria-selected": ariaSelected,
      onSelect,
      ...props
    } = t0);
    $[0] = t0;
    $[1] = ariaSelected;
    $[2] = onSelect;
    $[3] = props;
  } else {
    ariaSelected = $[1];
    onSelect = $[2];
    props = $[3];
  }
  let t1;
  if ($[4] !== onSelect) {
    t1 = event => {
      if (!event.defaultPrevented && typeof onSelect === "function") {
        onSelect(event);
      }
    };
    $[4] = onSelect;
    $[5] = t1;
  } else {
    t1 = $[5];
  }
  const clickHandler = t1;
  let t2;
  if ($[6] !== onSelect) {
    t2 = event_0 => {
      if ((event_0.key === " " || event_0.key === "Enter") && !event_0.defaultPrevented && typeof onSelect === "function") {
        onSelect(event_0);
      }
    };
    $[6] = onSelect;
    $[7] = t2;
  } else {
    t2 = $[7];
  }
  const keyDownHandler = t2;
  const t3 = ariaSelected ? 0 : -1;
  let t4;
  if ($[8] !== ariaSelected || $[9] !== clickHandler || $[10] !== keyDownHandler || $[11] !== props || $[12] !== t3) {
    t4 = /*#__PURE__*/jsxRuntime.jsx(UnderlineTabbedInterface.UnderlineItem, {
      as: "button",
      role: "tab",
      tabIndex: t3,
      "aria-selected": ariaSelected,
      type: "button",
      onClick: clickHandler,
      onKeyDown: keyDownHandler,
      ...props
    });
    $[8] = ariaSelected;
    $[9] = clickHandler;
    $[10] = keyDownHandler;
    $[11] = props;
    $[12] = t3;
    $[13] = t4;
  } else {
    t4 = $[13];
  }
  return t4;
};
Tab.displayName = 'UnderlinePanels.Tab';
const Panel = props => {
  const $ = reactCompilerRuntime.c(2);
  let t0;
  if ($[0] !== props) {
    t0 = /*#__PURE__*/jsxRuntime.jsx(BoxWithFallback.BoxWithFallback, {
      as: "div",
      role: "tabpanel",
      ...props
    });
    $[0] = props;
    $[1] = t0;
  } else {
    t0 = $[1];
  }
  return t0;
};
Panel.displayName = 'UnderlinePanels.Panel';
var UnderlinePanels$1 = Object.assign(UnderlinePanels, {
  Panel,
  Tab
});

module.exports = UnderlinePanels$1;
