// @sprucelabs/mercury-types/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateEventContract = exports.buildPermissionReference = exports.buildPermissionContract = exports.buildEventContract = exports.feedOptionsSchema = exports.permissionReferenceSchema = exports.statusFlagsSchema = exports.permissionContractSchema = exports.permissionSchema = exports.eventSignaturesByNameSchema = exports.eventSignatureSchema = exports.eventContractSchema = exports.permissionDefaultsSchema = void 0;
var permissionDefaults_schema_1 = require("./.spruce/schemas/mercury/v2020_12_25/permissionDefaults.schema");
Object.defineProperty(exports, "permissionDefaultsSchema", { enumerable: true, get: function () { return __importDefault(permissionDefaults_schema_1).default; } });
var eventContract_schema_1 = require("./.spruce/schemas/mercury/v2020_12_25/eventContract.schema");
Object.defineProperty(exports, "eventContractSchema", { enumerable: true, get: function () { return __importDefault(eventContract_schema_1).default; } });
var eventSignature_schema_1 = require("./.spruce/schemas/mercury/v2020_12_25/eventSignature.schema");
Object.defineProperty(exports, "eventSignatureSchema", { enumerable: true, get: function () { return __importDefault(eventSignature_schema_1).default; } });
var eventSignaturesByName_schema_1 = require("./.spruce/schemas/mercury/v2020_12_25/eventSignaturesByName.schema");
Object.defineProperty(exports, "eventSignaturesByNameSchema", { enumerable: true, get: function () { return __importDefault(eventSignaturesByName_schema_1).default; } });
var permission_schema_1 = require("./.spruce/schemas/mercury/v2020_12_25/permission.schema");
Object.defineProperty(exports, "permissionSchema", { enumerable: true, get: function () { return __importDefault(permission_schema_1).default; } });
var permissionContract_schema_1 = require("./.spruce/schemas/mercury/v2020_12_25/permissionContract.schema");
Object.defineProperty(exports, "permissionContractSchema", { enumerable: true, get: function () { return __importDefault(permissionContract_schema_1).default; } });
var statusFlags_schema_1 = require("./.spruce/schemas/mercury/v2020_12_25/statusFlags.schema");
Object.defineProperty(exports, "statusFlagsSchema", { enumerable: true, get: function () { return __importDefault(statusFlags_schema_1).default; } });
var permissionReference_schema_1 = require("./.spruce/schemas/mercury/v2020_12_25/permissionReference.schema");
Object.defineProperty(exports, "permissionReferenceSchema", { enumerable: true, get: function () { return __importDefault(permissionReference_schema_1).default; } });
var feedOptions_schema_1 = require("./.spruce/schemas/mercury/v2020_12_25/feedOptions.schema");
Object.defineProperty(exports, "feedOptionsSchema", { enumerable: true, get: function () { return __importDefault(feedOptions_schema_1).default; } });
__exportStar(require("./types/mercury.types"), exports);
__exportStar(require("./.spruce/schemas/schemas.types"), exports);
__exportStar(require("./constants"), exports);
var buildEventContract_1 = require("./utilities/buildEventContract");
Object.defineProperty(exports, "buildEventContract", { enumerable: true, get: function () { return __importDefault(buildEventContract_1).default; } });
var buildPermissionContract_1 = require("./utilities/buildPermissionContract");
Object.defineProperty(exports, "buildPermissionContract", { enumerable: true, get: function () { return __importDefault(buildPermissionContract_1).default; } });
var buildPermissionReference_1 = require("./utilities/buildPermissionReference");
Object.defineProperty(exports, "buildPermissionReference", { enumerable: true, get: function () { return __importDefault(buildPermissionReference_1).default; } });
var validateEventContract_1 = require("./utilities/validateEventContract");
Object.defineProperty(exports, "validateEventContract", { enumerable: true, get: function () { return __importDefault(validateEventContract_1).default; } });
